---
title: "Example analysis"
output: html_document
subtitle: "*ZSL Foster and NJ Gr&uuml;nwald*"
bibliography: "bibtexlib.bib"
---

```{r home_setup, echo=FALSE, warning=FALSE, message=FALSE}
options(width = 90)
set.seed(1)
# Knitr
library(knitr)
library(grid)
opts_chunk$set(dev = 'png', fig.width = 10, fig.height = 10, warning = TRUE,
               message = TRUE)
```

This is a short example analysis to give you a feel for the rest of the workshop.
If something does not make sense now, don't worry!
We will cover everything shown here in greater detail later.

## Reading data

The first step in any analysis is getting your data into R.
This can be difficult for taxonomic data since it has a hierarchical component (i.e. the taxonomic tree).
`Metacoder` [@foster2017metacoder] has functions for parsing (i.e. "reading") specific file formats used in metagenomics research.
However, for this demonstration, we will be using a more all-purpose parser from the `taxa` package [@foster2018taxa] meant for tabular data.

Included in `metacoder` is an example dataset that is a subset of the Human Microbiome Project data.
This dataset has two parts: 

* An abundance matrix called `hmp_otus`, with **samples in columns** and **OTUs in rows**
* A sample information table called `hmp_samples`, with **samples as rows** and columns of information describing the samples (e.g. gender)

```{r echo=FALSE}
knitr::include_graphics('./figure_sources/preferred_data_format_counts.png')
```

```{r echo=FALSE}
knitr::include_graphics('./figure_sources/preferred_data_format_samples.png')
```

This is a typical example of this kind of data and is the preferred way for R packages like `metacoder` and `taxa`.
Let us take a look at the dataset we will use in this example (just cut and paste the commands in the grey boxes into your Rstudio console to execute them): 

```{r message=FALSE}
library(metacoder)
print(hmp_otus)
print(hmp_samples)
```

We can now read the taxonomic information in the abundance matrix using a parser from `taxa`:

```{r}
obj <- parse_tax_data(hmp_otus,
                      class_cols = "lineage", # the column that contains taxonomic information
                      class_sep = ";", # The character used to separate taxa in the classification
                      class_regex = "^(.+)__(.+)$", # Regex identifying where the data for each taxon is
                      class_key = c(tax_rank = "info", # A key describing each regex capture group
                                    tax_name = "taxon_name"))
```

A parser grabs and translates information in alphanumeric format following some specification and breaks it into smaller elements available as a data structure for further analysis 
This command returns a `taxmap` data object.
The `taxmap` class is designed to store any number of tables, lists, or vectors associated with taxonomic information and facilitate manipulation of the data in a cohesive way.
Here is what that object looks like:

```{r}
print(obj)
```

Note how the original abundance matrix is contained in the `taxmap` object with an additional column called "taxon_id".
This table is stored in the list `obj$data`, which can contain any number of user-defined datasets.

## Abundance matrix manipulations

### Removing low-abundance counts

Low-abundance sequences might be the result of sequencing error, so typically we remove any counts/OTUs with less than some number of reads.
Lets set all counts with less than 5 reads to zero, overwriting the original table:

```{r}
obj$data$tax_data <- zero_low_counts(obj, dataset = "tax_data", min_count = 5)
```

There might now be some OTUs with no "real" reads. Lets check:

```{r}
no_reads <- rowSums(obj$data$tax_data[, hmp_samples$sample_id]) == 0
sum(no_reads)
```

It appears that `r sum(no_reads)`  of `r nrow(obj$data$tax_data)` OTUs now have no reads.
We can remove those OTUs and their associated taxa with `filter_obs`:

```{r}
obj <- filter_obs(obj, target = "tax_data", ! no_reads, drop_taxa = TRUE)
print(obj)
```

Note how there are fewer taxa now (`r length(taxon_names(obj))` from 174), as well as fewer OTUs (`r nrow(obj$data$tax_data)` from 1000).
This is because the `drop_taxa = TRUE` option caused any taxa without OTUs assigned to them after the filtering to be removed.
This coordinated manipulation of taxonomic and abundance data is one of the main benefits of using the `taxmap` class.


### Accounting for un-even sampling

These are raw counts, but people typically work with rarefied counts or proportions to avoid sampling depth biasing the results.
The function `rarefy_obs` will return the rarefied counts for a table in a taxmap object, but we will use proportions for this demonstration:

```{r}
obj$data$tax_data <- calc_obs_props(obj, "tax_data")
print(obj)
```


### Getting per-taxon information

Currently, we have values for the abundance of each OTU, not each taxon.
To get information on the taxa, we can sum the abundance per-taxon and add the results to the `taxmap` object in a new table:

```{r}
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = hmp_samples$sample_id)
```

Note that there is now an additional table called `tax_abund` with one row per taxon.
The name of the table is arbitrary; we could have called it anything.

```{r}
print(obj)
```

We can also easily calculate the number of samples that have reads for each taxon:

```{r}
obj$data$tax_occ <- calc_n_samples(obj, "tax_abund", groups = hmp_samples$body_site)
print(obj)
```

### Plotting taxonomic data

Now that we have per-taxon information (the `tax_abund` and `tax_occ` tables), we can plot the information using **heat trees** [@foster2017metacoder].
Heat trees are what we call taxonomic trees in which the node/branch size and color of a tree  correspond to some statistic of interest in analogy to heatmaps.
The code below plots the number of "Nose" microbiome samples that have reads for each taxon as the size of each taxon.
It also plots the number of OTUs assigned to each taxon in the overall dataset as color.

```{r}
heat_tree(obj, 
          node_label = taxon_names,
          node_size = n_obs,
          node_color = Nose, 
          node_size_axis_label = "OTU count",
          node_color_axis_label = "Samples with reads",
          layout = "davidson-harel", # The primary layout algorithm
          initial_layout = "reingold-tilford") # The layout algorithm that initializes node locations
```

> Be patient. This visualization takes a little bit of time to compute.

Note how we did not have to specify the full path to the variable "Nose", but just its name.
This is a shorthand for convenience.
We could have made the exact same plot using this command:

```{r, eval = FALSE}
heat_tree(obj, 
          node_label = obj$taxon_names(),
          node_size = obj$n_obs(),
          node_color = obj$data$tax_occ$Nose, 
          node_size_axis_label = "OTU count",
          node_color_axis_label = "Samples with reads",
          layout = "davidson-harel", # The primary layout algorithm
          initial_layout = "reingold-tilford") # The layout algorithm that initializes node locations
```

This is known as non-standard evaluation in programmer jargon and will be used in many functions throughout this workshop.

### Alpha diversity

Alpha diversity is a measure of the diversity **within** each sample or group of samples.
It can be calculated at any rank of the taxonomy, but it is usually calculated at the species or OTU "rank".
There are multiple methods used to calculate a value to represent alpha diversity. 
The simplest is just the number of species, but the ones used most often factor in how common each species is as well.
Below, we calculate the alpha diversity of OTUs using the inverse Simpson's index [@hill1973diversity; @simpson1949measurement] using the package `vegan` [@oksanen2007vegan]. 

```{r}
hmp_samples$inv_simp <- vegan::diversity(obj$data$tax_data[, hmp_samples$sample_id],
                                         index = "invsimpson",
                                         MARGIN = 2) # What orietation the matrix is in
```

Adding this on to the sample data makes it easy to use the sample information in graphing. 
Lets compare the alpha diversity of samples from males and females using `ggplot2`, a popular R package for plotting. 

```{r fig.height=5, fig.width=5}
library(ggplot2)
ggplot(hmp_samples, aes(x = sex, y = inv_simp)) +
  geom_boxplot()
```

Not much difference there, as you might expect. 
We can also compare body sites:

```{r fig.height=5, fig.width=5}
ggplot(hmp_samples, aes(x = body_site, y = inv_simp)) +
  geom_boxplot()
```

That's more interesting; skin has much lower diversity than any of the wetter areas, which makes sense.
Lets see if that's a significant difference using an ANOVA.

```{r}
anova_result <- aov(inv_simp ~ body_site, hmp_samples)
summary(anova_result)
```

That tells that at least one of the body site means is different from the other, but not which one (although we can make a good guess).
A Tukey's HSD test can compare each site to every other and tell us which are significantly different.
Although base R has a Tukey's HSD function called `TukeyHSD`, we will use one from the package `agricolae` since it supplies grouping codes that are useful for graphing.

```{r}
library(agricolae)
tukey_result <- HSD.test(anova_result, "body_site", group = TRUE)
print(tukey_result)
```

We are interested in the `$groups` table that says which sites are different.
With a little tweaking, we can add this data to the graph we made.
Lets add some nicer text as well.

```{r fig.height=5, fig.width=5}
group_data <- tukey_result$groups[order(rownames(tukey_result$groups)),]
ggplot(hmp_samples, aes(x = body_site, y = inv_simp)) +
  geom_text(data = data.frame(),
            aes(x = rownames(group_data), y = max(hmp_samples$inv_simp) + 1, label = group_data$groups),
            col = 'black',
            size = 10) +
  geom_boxplot() +
  ggtitle("Alpha diversity of human body microbiomes") +
  xlab("Body site") +
  ylab("Inverse Simpson's index")
```

This tells us that microbiomes from saliva and skin are significantly different from each other, but not significantly different from any other community.

### Comparing two treatments/groups

Usually we are interested in how groups of samples compare.
For example, we might want to know which taxa differ between the nose and throat, or between men and women.
The function `compare_groups` facilitates these comparisons:

```{r, warning = FALSE}
obj$data$diff_table <- compare_groups(obj,
                                      dataset = "tax_abund",
                                      cols = hmp_samples$sample_id, # What columns of sample data to use
                                      groups = hmp_samples$sex) # What category each sample is assigned to
print(obj$data$diff_table)
```

We can use this information to create what we call a **differential heat tree** [@foster2017metacoder], which indicates which taxa are more abundant in each treatment: 

```{r}
heat_tree(obj, 
          node_label = taxon_names,
          node_size = n_obs, # n_obs is a function that calculates, in this case, the number of OTUs per taxon
          node_color = log2_median_ratio, # A column from `obj$data$diff_table`
          node_color_interval = c(-2, 2), # The range of `log2_median_ratio` to display
          node_color_range = c("cyan", "gray", "tan"), # The color palette used
          node_size_axis_label = "OTU count",
          node_color_axis_label = "Log 2 ratio of median proportions",
          layout = "davidson-harel", # The primary layout algorithm
          initial_layout = "reingold-tilford") # The layout algorithm that initializes node locations
```

In this case, taxa colored tan are more abundant in women and those colored blue are more abundant in men.
Note that we have not taken into account statistical significance, so lets include that, too.
First, we need to correct for multiple comparisons: 

```{r}
obj$data$diff_table$wilcox_p_value <- p.adjust(obj$data$diff_table$wilcox_p_value,
                                               method = "fdr")
```

If we then look at the distribution of p-values, we can see that none are even close to significant:

```{r}
range(obj$data$diff_table$wilcox_p_value, finite = TRUE) 
```

There is no need to graph this, but if there still were some significant differences, we could set any difference that is not significant to zero and repeat the last `heat_tree` command.

### Comparing any number of treatments/groups

A single differential heat tree can compare two treatments only (e.g., male and female), but what if you have more treatments?
In that case we can make a matrix of heat trees, one for each pairwise comparison of treatments as follows: 

```{r, warning = FALSE}
obj$data$diff_table <- compare_groups(obj, dataset = "tax_abund",
                                      cols = hmp_samples$sample_id, # What columns of sample data to use
                                      groups = hmp_samples$body_site) # What category each sample is assigned to
print(obj$data$diff_table)
```

There is a special function to plot this type of data called `heat_tree_matrix` [@foster2017metacoder]:

```{r}
heat_tree_matrix(obj,
                 dataset = "diff_table",
                 node_size = n_obs, # n_obs is a function that calculates, in this case, the number of OTUs per taxon
                 node_label = taxon_names,
                 node_color = log2_median_ratio, # A column from `obj$data$diff_table`
                 node_color_range = diverging_palette(), # The built-in palette for diverging data
                 node_color_trans = "linear", # The default is scaled by circle area
                 node_color_interval = c(-3, 3), # The range of `log2_median_ratio` to display
                 edge_color_interval = c(-3, 3), # The range of `log2_median_ratio` to display
                 node_size_axis_label = "Number of OTUs",
                 node_color_axis_label = "Log2 ratio median proportions",
                 layout = "davidson-harel", # The primary layout algorithm
                 initial_layout = "reingold-tilford", # The layout algorithm that initializes node locations
                 output_file = "differential_heat_tree.pdf") # Saves the plot as a pdf file
```

The grey tree on the lower left functions as a key for the unlabeled trees.
Each of the smaller trees represent a comparison between body sites in the columns and rows.
A taxon colored brown is more abundant in the body site in the column and a taxon colored green is more abundant in body site of the row.
For example, *Bacteroidetes* are more abundant (i.e., have more reads) in the throat samples than the nose samples and this is  due to the greater abundance of two genera within this group: *Prevotella* and *Porphyromonas*.
Download and open the PDF file to explore the details of the plot zooming in on any branches of interest 

```{r, child="_sessioninfo.Rmd"}
```

```{r cleanup, message=FALSE, warning=FALSE, include=FALSE}
file.remove("differential_heat_tree.pdf")
```

References
----------

<!---- Reference list ---->
