<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Example analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-52691601-4"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-52691601-4');
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 61px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 66px;
  margin-top: -66px;
}

.section h2 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h3 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h4 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h5 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h6 {
  padding-top: 66px;
  margin-top: -66px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Analysis of Microbiome Data in R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="02--quick_example.html">Example analysis</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00--required_software.html">Required software</a>
    </li>
    <li>
      <a href="00--required_data.html">Required datasets</a>
    </li>
    <li>
      <a href="03--parsing.html">Getting data into R</a>
    </li>
    <li>
      <a href="04--manipulating.html">Manipulating taxonomic data</a>
    </li>
    <li>
      <a href="05--plotting.html">Plotting taxonomic data</a>
    </li>
    <li>
      <a href="06--quality_control.html">Data quality control</a>
    </li>
    <li>
      <a href="07--diversity_stats.html">Diversity and Ordination</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Appendices
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00--intro_to_r.html">Introduction to R</a>
    </li>
    <li>
      <a href="00--intro_to_rstudio.html">Introduction to RStudio</a>
    </li>
    <li>
      <a href="00--glossary.html">Glossary</a>
    </li>
    <li>
      <a href="https://github.com/grunwaldlab/analysis_of_microbiome_community_data_in_r">Website source code</a>
    </li>
  </ul>
</li>
<li>
  <a href="about.html">
    <span class="fa fa-info"></span>
     
    About
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Example analysis</h1>

</div>


<p>This is a short example analysis to give you a feel for how <code>metacoder</code> and <code>taxa</code> are used in microbiome analysis. If something does not make sense now, don’t worry! We will cover everything shown here in greater detail later.</p>
<div id="reading-data" class="section level2">
<h2>Reading data</h2>
<p>The first step in any analysis is getting your data into R. This can be difficult for taxonomic data since it has a hierarchical component (i.e., the taxonomic tree). <code>Metacoder</code> has functions for <b><a href ="00--glossary.html#parsing_anchor">parsing</a></b> specific file formats used in metagenomics research. However, for this demonstration, we will be using a more all-purpose parser from the <code>taxa</code> package meant for tabular data.</p>
<p>Included in <code>metacoder</code> is an example dataset that is a subset of the Human Microbiome Project data. This dataset has two parts:</p>
<ul>
<li>An abundance matrix called <code>hmp_otus</code>, with <strong>samples in columns</strong> and <b><a href ="00--glossary.html#operational_taxonomic_units_(otus)_anchor">Operational Taxonomic Units (OTUs)</a></b> in rows</li>
<li>A sample data table called <code>hmp_samples</code>, with <strong>samples as rows</strong> and columns of information describing the samples (e.g., gender)</li>
</ul>
<p>This is a typical way for this kind of data to be formatted and is the preferred way for packages like <code>metacoder</code> and <code>taxa</code>. Lets take a look at the dataset we will use in this example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(metacoder)
<span class="kw">print</span>(hmp_otus)</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 1,000 x 52</span><span>
##    otu_id  lineage          `700035949` `700097855` `700100489` `700111314` `700033744` `700109581`
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                  </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> OTU_97… r__Root;p__Prot…           0           2           1           0           0           0
## </span><span style='color: #BCBCBC;'> 2</span><span> OTU_97… r__Root;p__Bact…           0           0           0           0           0           0
## </span><span style='color: #BCBCBC;'> 3</span><span> OTU_97… r__Root;p__Bact…           0           1           0           0           0           0
## </span><span style='color: #BCBCBC;'> 4</span><span> OTU_97… r__Root;p__Acti…           8          36          10           5          66          38
## </span><span style='color: #BCBCBC;'> 5</span><span> OTU_97… r__Root;p__Prot…           3          25           0           0           0           1
## </span><span style='color: #BCBCBC;'> 6</span><span> OTU_97… r__Root;p__Acti…          42         277          16          22          85         211
## </span><span style='color: #BCBCBC;'> 7</span><span> OTU_97… r__Root;p__Firm…           4          17          21           1          74          12
## </span><span style='color: #BCBCBC;'> 8</span><span> OTU_97… r__Root;p__Bact…           0           0           0           0           0           0
## </span><span style='color: #BCBCBC;'> 9</span><span> OTU_97… r__Root;p__Firm…           0           0           0           0           0           0
## </span><span style='color: #BCBCBC;'>10</span><span> OTU_97… r__Root;p__Bact…           0           0           0           0           1           0
## </span><span style='color: #949494;'># ... with 990 more rows, and 44 more variables: `700111044` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>, `700101365` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>,
## #   `700100431` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>, `700016050` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>, `700032425` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>, `700024855` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>, `700103488` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>,
## #   `700096869` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>, `700107379` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>, `700096422` </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span style='color: #949494;'>, …</span><span>
</span></code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(hmp_samples)</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 50 x 3</span><span>
## </span><span style='color: #949494;'># Groups:   body_site, sex [10]</span><span>
##    sample_id sex    body_site
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    
## </span><span style='color: #BCBCBC;'> 1</span><span> 700035949 female Nose     
## </span><span style='color: #BCBCBC;'> 2</span><span> 700097855 female Nose     
## </span><span style='color: #BCBCBC;'> 3</span><span> 700100489 female Nose     
## </span><span style='color: #BCBCBC;'> 4</span><span> 700111314 female Nose     
## </span><span style='color: #BCBCBC;'> 5</span><span> 700033744 female Nose     
## </span><span style='color: #BCBCBC;'> 6</span><span> 700109581 male   Nose     
## </span><span style='color: #BCBCBC;'> 7</span><span> 700111044 male   Nose     
## </span><span style='color: #BCBCBC;'> 8</span><span> 700101365 male   Nose     
## </span><span style='color: #BCBCBC;'> 9</span><span> 700100431 male   Nose     
## </span><span style='color: #BCBCBC;'>10</span><span> 700016050 male   Nose     
## </span><span style='color: #949494;'># ... with 40 more rows</span><span>
</span></code></pre>
<p>One challenge this data presents is the format of the taxonomic information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hmp_otus<span class="op">$</span>lineage[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]</code></pre></div>
<pre class="r-output"><code>## [1] "r__Root;p__Proteobacteria;c__Gammaproteobacteria;o__Pasteurellales;f__Pasteurellaceae;g__Haemophilus"       
## [2] "r__Root;p__Bacteroidetes;c__Flavobacteria;o__Flavobacteriales;f__Flavobacteriaceae;g__Capnocytophaga"       
## [3] "r__Root;p__Bacteroidetes;c__Bacteroidia;o__Bacteroidales;f__Porphyromonadaceae;g__Porphyromonas"            
## [4] "r__Root;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Propionibacteriaceae;g__Propionibacterium"
</code></pre>
<p>We can process the abundance matrix, and parse the taxonomic information at the same time, using a parser from <code>taxa</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj &lt;-<span class="st"> </span><span class="kw">parse_tax_data</span>(hmp_otus,
                      <span class="dt">class_cols =</span> <span class="st">&quot;lineage&quot;</span>, <span class="co"># the column that contains taxonomic information</span>
                      <span class="dt">class_sep =</span> <span class="st">&quot;;&quot;</span>, <span class="co"># The character used to separate taxa in the classification</span>
                      <span class="dt">class_regex =</span> <span class="st">&quot;^(.+)__(.+)$&quot;</span>, <span class="co"># Regex identifying where the data for each taxon is</span>
                      <span class="dt">class_key =</span> <span class="kw">c</span>(<span class="dt">tax_rank =</span> <span class="st">&quot;info&quot;</span>, <span class="co"># A key describing each regex capture group</span>
                                    <span class="dt">tax_name =</span> <span class="st">&quot;taxon_name&quot;</span>))</code></pre></div>
<p>This returns a <code>taxmap</code> object. The <code>taxmap</code> class is designed to store any number of tables, <b><a href ="00--glossary.html#list_anchor">lists</a></b>, or <b><a href ="00--glossary.html#vector_anchor">vectors</a></b> associated with taxonomic information and facilitate manipulating the data. Here is what that object looks like:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   174 taxa: <span style='color: #00BB00;'>ab</span><span>. Root, </span><span style='color: #00BB00;'>ac</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>gr</span><span>. Blautia, </span><span style='color: #00BB00;'>gs</span><span>. Clostridium
##   174 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>ab</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ac</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ad</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ae</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>af</span><span> ... </span><span style='color: #00BB00;'>dk</span><span>-&gt;</span><span style='color: #00BB00;'>gp</span><span>, </span><span style='color: #00BB00;'>cm</span><span>-&gt;</span><span style='color: #00BB00;'>gq</span><span>, </span><span style='color: #00BB00;'>cf</span><span>-&gt;</span><span style='color: #00BB00;'>gr</span><span>, </span><span style='color: #00BB00;'>cw</span><span>-&gt;</span><span style='color: #00BB00;'>gs</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 1,000 x 53</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>otu_id  lineage          `700035949` `700097855` `700100489` `700111314`
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;                  &lt;int&gt;       &lt;int&gt;       &lt;int&gt;       &lt;int&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>dm</span><span>       OTU_97… r__Root;p__Prot…           0           2           1           0
##       2 </span><span style='color: #00BB00;'>dn</span><span>       OTU_97… r__Root;p__Bact…           0           0           0           0
##       3 </span><span style='color: #00BB00;'>do</span><span>       OTU_97… r__Root;p__Bact…           0           1           0           0
##       # ... with 997 more rows, and 46 more variables: `700033744` &lt;int&gt;,
##       #   `700109581` &lt;int&gt;, `700111044` &lt;int&gt;, `700101365` &lt;int&gt;, `700100431` &lt;int&gt;,
##       #   `700016050` &lt;int&gt;, `700032425` &lt;int&gt;, `700024855` &lt;int&gt;, `700103488` &lt;int&gt;,
##       #   `700096869` &lt;int&gt;, …
##     </span><span style='font-weight: bold;'>class_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 5,922 x 5</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>input_index tax_rank tax_name            regex_match           
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;               &lt;chr&gt;                 </span><span>
##       1 </span><span style='color: #00BB00;'>ab</span><span>                 1 r        Root                r__Root               
##       2 </span><span style='color: #00BB00;'>ac</span><span>                 1 p        Proteobacteria      p__Proteobacteria     
##       3 </span><span style='color: #00BB00;'>aj</span><span>                 1 c        Gammaproteobacteria c__Gammaproteobacteria
##       # ... with 5,919 more rows
##   0 functions:
</span></code></pre>
<p>Note how the original abundance matrix is contained in the <code>taxmap</code> <b><a href ="00--glossary.html#object_anchor">object</a></b> with an additional column called “taxon_id”. This table is stored in the list <code>obj$data</code>, which can contain any number of user-defined datasets. Also note that we have a different number of taxa and OTUs. This is different from a traditional ecological dataset. An OTU may contain sequence variants, so a single OTU represents multiple similar sequences. These OTUs are assigned to taxa so a taxon may include multiple OTUs. Taxa may be organized in different <b><a href ="00--glossary.html#taxonomic_ranks_anchor">ranks</a></b>. For example, one taxon might be a species while another might be a genus. These abstractions are necessary because we may have sequences that we can not confidently assign to a traditional taxon.</p>
</div>
<div id="abundance-matrix-manipulations" class="section level2">
<h2>Abundance matrix manipulations</h2>
<div id="removing-low-abundance-counts" class="section level3">
<h3>Removing low-abundance counts</h3>
<p>Recall that the abundance matrix contains samples in columns and OTUs in rows. Each cell is the number of times an OTU was observed in a sample. Some of these cells may contain a low number of observations. These low-abundance sequences might be the result of sequencing error, so typically we remove any counts/OTUs with less than some number of reads. Lets set all counts with less than 5 reads to zero, overwriting the original table:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>tax_data &lt;-<span class="st"> </span><span class="kw">zero_low_counts</span>(obj, <span class="dt">dataset =</span> <span class="st">&quot;tax_data&quot;</span>, <span class="dt">min_count =</span> <span class="dv">5</span>)</code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    700035949, 700097855, 700100489, 700111314 ... 700095535, 700102367, 700101358</code></pre>
<pre><code>## Zeroing 4325 of 50000 counts less than 5.</code></pre>
<p>By setting low abundance counts to zero we might have created OTUs that no longer contain any observations. We can check as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">no_reads &lt;-<span class="st"> </span><span class="kw">rowSums</span>(obj<span class="op">$</span>data<span class="op">$</span>tax_data[, hmp_samples<span class="op">$</span>sample_id]) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>
<span class="kw">sum</span>(no_reads)</code></pre></div>
<pre class="r-output"><code>## [1] 211
</code></pre>
<p>It appears that 211 of 1000 OTUs now have no reads. We can remove those OTUs and their associated taxa with <code>filter_obs</code> from the <code>taxa</code> package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj &lt;-<span class="st"> </span><span class="kw">filter_obs</span>(obj, <span class="dt">target =</span> <span class="st">&quot;tax_data&quot;</span>, <span class="op">!</span><span class="st"> </span>no_reads, <span class="dt">drop_taxa =</span> <span class="ot">TRUE</span>)
<span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   155 taxa: <span style='color: #00BB00;'>ab</span><span>. Root, </span><span style='color: #00BB00;'>ac</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>gq</span><span>. Collinsella, </span><span style='color: #00BB00;'>gs</span><span>. Clostridium
##   155 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>ab</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ac</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ad</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ae</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>af</span><span> ... </span><span style='color: #00BB00;'>dj</span><span>-&gt;</span><span style='color: #00BB00;'>gn</span><span>, </span><span style='color: #00BB00;'>dk</span><span>-&gt;</span><span style='color: #00BB00;'>gp</span><span>, </span><span style='color: #00BB00;'>cm</span><span>-&gt;</span><span style='color: #00BB00;'>gq</span><span>, </span><span style='color: #00BB00;'>cw</span><span>-&gt;</span><span style='color: #00BB00;'>gs</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 789 x 51</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>`700035949` `700097855` `700100489` `700111314` `700033744` `700109581`
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>dm</span><span>                0.          0.          0.          0.          0.          0.
##       2 </span><span style='color: #00BB00;'>dn</span><span>                0.          0.          0.          0.          0.          0.
##       3 </span><span style='color: #00BB00;'>do</span><span>                0.          0.          0.          0.          0.          0.
##       # ... with 786 more rows, and 44 more variables: `700111044` &lt;dbl&gt;,
##       #   `700101365` &lt;dbl&gt;, `700100431` &lt;dbl&gt;, `700016050` &lt;dbl&gt;, `700032425` &lt;dbl&gt;,
##       #   `700024855` &lt;dbl&gt;, `700103488` &lt;dbl&gt;, `700096869` &lt;dbl&gt;, `700107379` &lt;dbl&gt;,
##       #   `700096422` &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>class_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 5,903 x 5</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>input_index tax_rank tax_name            regex_match           
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;               &lt;chr&gt;                 </span><span>
##       1 </span><span style='color: #00BB00;'>ab</span><span>                 1 r        Root                r__Root               
##       2 </span><span style='color: #00BB00;'>ac</span><span>                 1 p        Proteobacteria      p__Proteobacteria     
##       3 </span><span style='color: #00BB00;'>aj</span><span>                 1 c        Gammaproteobacteria c__Gammaproteobacteria
##       # ... with 5,900 more rows
##   0 functions:
</span></code></pre>
<p>Note how there are fewer taxa now (155 from 174), as well as fewer OTUs (789 from 1000). This is because the <code>drop_taxa = TRUE</code> option caused any taxa without OTUs assigned to them after the filtering to be removed. This coordinated manipulation of taxonomic and abundance data is one of the main benefits of using the <code>taxmap</code> class.</p>
</div>
<div id="accounting-for-un-even-sampling" class="section level3">
<h3>Accounting for un-even sampling</h3>
<p>Ideally, we would sequence each sample the same amount (i.e., the same number of reads). However, sequencing technologies are imperfect, so some samples may have more reads than others. This creates a situation where we may have observed more diversity in some samples because they were sequenced more thoroughly than others. So far we’ve used raw counts, but people typically work with <b><a href ="00--glossary.html#rarefaction_anchor">rarefied</a></b> counts or proportions to try to avoid the possibility of sampling depth biasing the results. Here we use the function <code>calc_obs_props</code> to divide each sample’s counts by the total number of counts observed for each sample, resulting in a proportion.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>tax_data &lt;-<span class="st"> </span><span class="kw">calc_obs_props</span>(obj, <span class="st">&quot;tax_data&quot;</span>)</code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    700035949, 700097855, 700100489, 700111314 ... 700095535, 700102367, 700101358</code></pre>
<pre><code>## Calculating proportions from counts for 50 columns for 789 observations.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   155 taxa: <span style='color: #00BB00;'>ab</span><span>. Root, </span><span style='color: #00BB00;'>ac</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>gq</span><span>. Collinsella, </span><span style='color: #00BB00;'>gs</span><span>. Clostridium
##   155 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>ab</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ac</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ad</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ae</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>af</span><span> ... </span><span style='color: #00BB00;'>dj</span><span>-&gt;</span><span style='color: #00BB00;'>gn</span><span>, </span><span style='color: #00BB00;'>dk</span><span>-&gt;</span><span style='color: #00BB00;'>gp</span><span>, </span><span style='color: #00BB00;'>cm</span><span>-&gt;</span><span style='color: #00BB00;'>gq</span><span>, </span><span style='color: #00BB00;'>cw</span><span>-&gt;</span><span style='color: #00BB00;'>gs</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 789 x 51</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>`700035949` `700097855` `700100489` `700111314` `700033744` `700109581`
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>dm</span><span>                0.          0.          0.          0.          0.          0.
##       2 </span><span style='color: #00BB00;'>dn</span><span>                0.          0.          0.          0.          0.          0.
##       3 </span><span style='color: #00BB00;'>do</span><span>                0.          0.          0.          0.          0.          0.
##       # ... with 786 more rows, and 44 more variables: `700111044` &lt;dbl&gt;,
##       #   `700101365` &lt;dbl&gt;, `700100431` &lt;dbl&gt;, `700016050` &lt;dbl&gt;, `700032425` &lt;dbl&gt;,
##       #   `700024855` &lt;dbl&gt;, `700103488` &lt;dbl&gt;, `700096869` &lt;dbl&gt;, `700107379` &lt;dbl&gt;,
##       #   `700096422` &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>class_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 5,903 x 5</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>input_index tax_rank tax_name            regex_match           
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;               &lt;chr&gt;                 </span><span>
##       1 </span><span style='color: #00BB00;'>ab</span><span>                 1 r        Root                r__Root               
##       2 </span><span style='color: #00BB00;'>ac</span><span>                 1 p        Proteobacteria      p__Proteobacteria     
##       3 </span><span style='color: #00BB00;'>aj</span><span>                 1 c        Gammaproteobacteria c__Gammaproteobacteria
##       # ... with 5,900 more rows
##   0 functions:
</span></code></pre>
</div>
<div id="getting-per-taxon-information" class="section level3">
<h3>Getting per-taxon information</h3>
<p>Currently, we have values for the abundance of each OTU, not each taxon. To get information on the taxa, we can sum the abundance per-taxon and add the results to the <code>taxmap</code> object in a new table:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>tax_abund &lt;-<span class="st"> </span><span class="kw">calc_taxon_abund</span>(obj, <span class="st">&quot;tax_data&quot;</span>,
                                       <span class="dt">cols =</span> hmp_samples<span class="op">$</span>sample_id)</code></pre></div>
<pre><code>## Summing per-taxon counts from 50 columns for 155 taxa</code></pre>
<p>Note that there is now an additional table called <code>tax_abund</code> with one row per taxon. The name of the table is arbitrary; we could have called it anything.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   155 taxa: <span style='color: #00BB00;'>ab</span><span>. Root, </span><span style='color: #00BB00;'>ac</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>gq</span><span>. Collinsella, </span><span style='color: #00BB00;'>gs</span><span>. Clostridium
##   155 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>ab</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ac</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ad</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ae</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>af</span><span> ... </span><span style='color: #00BB00;'>dj</span><span>-&gt;</span><span style='color: #00BB00;'>gn</span><span>, </span><span style='color: #00BB00;'>dk</span><span>-&gt;</span><span style='color: #00BB00;'>gp</span><span>, </span><span style='color: #00BB00;'>cm</span><span>-&gt;</span><span style='color: #00BB00;'>gq</span><span>, </span><span style='color: #00BB00;'>cw</span><span>-&gt;</span><span style='color: #00BB00;'>gs</span><span>
##   3 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 789 x 51</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>`700035949` `700097855` `700100489` `700111314` `700033744` `700109581`
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>dm</span><span>                0.          0.          0.          0.          0.          0.
##       2 </span><span style='color: #00BB00;'>dn</span><span>                0.          0.          0.          0.          0.          0.
##       3 </span><span style='color: #00BB00;'>do</span><span>                0.          0.          0.          0.          0.          0.
##       # ... with 786 more rows, and 44 more variables: `700111044` &lt;dbl&gt;,
##       #   `700101365` &lt;dbl&gt;, `700100431` &lt;dbl&gt;, `700016050` &lt;dbl&gt;, `700032425` &lt;dbl&gt;,
##       #   `700024855` &lt;dbl&gt;, `700103488` &lt;dbl&gt;, `700096869` &lt;dbl&gt;, `700107379` &lt;dbl&gt;,
##       #   `700096422` &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>class_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 5,903 x 5</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>input_index tax_rank tax_name            regex_match           
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;               &lt;chr&gt;                 </span><span>
##       1 </span><span style='color: #00BB00;'>ab</span><span>                 1 r        Root                r__Root               
##       2 </span><span style='color: #00BB00;'>ac</span><span>                 1 p        Proteobacteria      p__Proteobacteria     
##       3 </span><span style='color: #00BB00;'>aj</span><span>                 1 c        Gammaproteobacteria c__Gammaproteobacteria
##       # ... with 5,900 more rows
##     </span><span style='font-weight: bold;'>tax_abund</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 155 x 51</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>`700035949` `700097855` `700100489` `700111314` `700033744` `700109581`
## </span><span style='font-style: italic;'>      * &lt;chr&gt;          &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ab</span><span>             1.00      1.00             1.      1.00         1.00      1.00   
##       2 </span><span style='color: #00BB00;'>ac</span><span>             0.206     0.0262           0.      0.252        0.225     0.00187
##       3 </span><span style='color: #00BB00;'>ad</span><span>             0.        0.00269          0.      0.0417       0.        0.     
##       # ... with 152 more rows, and 44 more variables: `700111044` &lt;dbl&gt;,
##       #   `700101365` &lt;dbl&gt;, `700100431` &lt;dbl&gt;, `700016050` &lt;dbl&gt;, `700032425` &lt;dbl&gt;,
##       #   `700024855` &lt;dbl&gt;, `700103488` &lt;dbl&gt;, `700096869` &lt;dbl&gt;, `700107379` &lt;dbl&gt;,
##       #   `700096422` &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
<p>We can also easily calculate the number of samples that have reads for each taxon:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>tax_occ &lt;-<span class="st"> </span><span class="kw">calc_n_samples</span>(obj, <span class="st">&quot;tax_abund&quot;</span>, <span class="dt">groups =</span> hmp_samples<span class="op">$</span>body_site)</code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    700035949, 700097855, 700100489, 700111314 ... 700095535, 700102367, 700101358</code></pre>
<pre><code>## Calculating number of samples with a value greater than 0 for 50 columns in 5 groups for 155 observations</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   155 taxa: <span style='color: #00BB00;'>ab</span><span>. Root, </span><span style='color: #00BB00;'>ac</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>gq</span><span>. Collinsella, </span><span style='color: #00BB00;'>gs</span><span>. Clostridium
##   155 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>ab</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ac</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ad</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>ae</span><span>, </span><span style='color: #00BB00;'>ab</span><span>-&gt;</span><span style='color: #00BB00;'>af</span><span> ... </span><span style='color: #00BB00;'>dj</span><span>-&gt;</span><span style='color: #00BB00;'>gn</span><span>, </span><span style='color: #00BB00;'>dk</span><span>-&gt;</span><span style='color: #00BB00;'>gp</span><span>, </span><span style='color: #00BB00;'>cm</span><span>-&gt;</span><span style='color: #00BB00;'>gq</span><span>, </span><span style='color: #00BB00;'>cw</span><span>-&gt;</span><span style='color: #00BB00;'>gs</span><span>
##   4 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 789 x 51</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>`700035949` `700097855` `700100489` `700111314` `700033744` `700109581`
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>dm</span><span>                0.          0.          0.          0.          0.          0.
##       2 </span><span style='color: #00BB00;'>dn</span><span>                0.          0.          0.          0.          0.          0.
##       3 </span><span style='color: #00BB00;'>do</span><span>                0.          0.          0.          0.          0.          0.
##       # ... with 786 more rows, and 44 more variables: `700111044` &lt;dbl&gt;,
##       #   `700101365` &lt;dbl&gt;, `700100431` &lt;dbl&gt;, `700016050` &lt;dbl&gt;, `700032425` &lt;dbl&gt;,
##       #   `700024855` &lt;dbl&gt;, `700103488` &lt;dbl&gt;, `700096869` &lt;dbl&gt;, `700107379` &lt;dbl&gt;,
##       #   `700096422` &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>class_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 5,903 x 5</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>input_index tax_rank tax_name            regex_match           
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;               &lt;chr&gt;                 </span><span>
##       1 </span><span style='color: #00BB00;'>ab</span><span>                 1 r        Root                r__Root               
##       2 </span><span style='color: #00BB00;'>ac</span><span>                 1 p        Proteobacteria      p__Proteobacteria     
##       3 </span><span style='color: #00BB00;'>aj</span><span>                 1 c        Gammaproteobacteria c__Gammaproteobacteria
##       # ... with 5,900 more rows
##     </span><span style='font-weight: bold;'>tax_abund</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 155 x 51</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>`700035949` `700097855` `700100489` `700111314` `700033744` `700109581`
## </span><span style='font-style: italic;'>      * &lt;chr&gt;          &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ab</span><span>             1.00      1.00             1.      1.00         1.00      1.00   
##       2 </span><span style='color: #00BB00;'>ac</span><span>             0.206     0.0262           0.      0.252        0.225     0.00187
##       3 </span><span style='color: #00BB00;'>ad</span><span>             0.        0.00269          0.      0.0417       0.        0.     
##       # ... with 152 more rows, and 44 more variables: `700111044` &lt;dbl&gt;,
##       #   `700101365` &lt;dbl&gt;, `700100431` &lt;dbl&gt;, `700016050` &lt;dbl&gt;, `700032425` &lt;dbl&gt;,
##       #   `700024855` &lt;dbl&gt;, `700103488` &lt;dbl&gt;, `700096869` &lt;dbl&gt;, `700107379` &lt;dbl&gt;,
##       #   `700096422` &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>tax_occ</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 155 x 6</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span> Nose Saliva  Skin Stool Throat
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;int&gt;  &lt;int&gt; &lt;int&gt; &lt;int&gt;  &lt;int&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ab</span><span>          10     10    10    10     10
##       2 </span><span style='color: #00BB00;'>ac</span><span>           8     10     9     2     10
##       3 </span><span style='color: #00BB00;'>ad</span><span>           5     10     8    10     10
##       # ... with 152 more rows
##   0 functions:
</span></code></pre>
</div>
<div id="plotting-taxonomic-data" class="section level3">
<h3>Plotting taxonomic data</h3>
<p>Now that we have per-taxon information (The <code>tax_abund</code> and <code>tax_occ</code> tables), we can plot the information using <strong>heat trees</strong>. Heat trees are what we call taxonomic trees in which the size and color of tree parts correspond to some statistic of interest. The code below plots the number of “Nose” samples that have reads for each taxon as the size of each taxon. It also plots the number of OTUs assigned to each taxon in the overall dataset as color.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>) <span class="co"># This makes the plot appear the same each time it is run </span>
<span class="kw">heat_tree</span>(obj, 
          <span class="dt">node_label =</span> taxon_names,
          <span class="dt">node_size =</span> n_obs,
          <span class="dt">node_color =</span> Nose, 
          <span class="dt">node_size_axis_label =</span> <span class="st">&quot;OTU count&quot;</span>,
          <span class="dt">node_color_axis_label =</span> <span class="st">&quot;Samples with reads&quot;</span>,
          <span class="dt">layout =</span> <span class="st">&quot;davidson-harel&quot;</span>, <span class="co"># The primary layout algorithm</span>
          <span class="dt">initial_layout =</span> <span class="st">&quot;reingold-tilford&quot;</span>) <span class="co"># The layout algorithm that initializes node locations</span></code></pre></div>
<p><img src="02--quick_example_files/figure-html/unnamed-chunk-15-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>Note how we did not have to specify the full path to the variable “Nose”, but just its name. This is a shorthand for convenience. We could have made the exact same plot using this command:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
<span class="kw">heat_tree</span>(obj, 
          <span class="dt">node_label =</span> obj<span class="op">$</span><span class="kw">taxon_names</span>(),
          <span class="dt">node_size =</span> obj<span class="op">$</span><span class="kw">n_obs</span>(),
          <span class="dt">node_color =</span> obj<span class="op">$</span>data<span class="op">$</span>tax_occ<span class="op">$</span>Nose, 
          <span class="dt">node_size_axis_label =</span> <span class="st">&quot;OTU count&quot;</span>,
          <span class="dt">node_color_axis_label =</span> <span class="st">&quot;Samples with reads&quot;</span>,
          <span class="dt">layout =</span> <span class="st">&quot;davidson-harel&quot;</span>, <span class="co"># The primary layout algorithm</span>
          <span class="dt">initial_layout =</span> <span class="st">&quot;reingold-tilford&quot;</span>) <span class="co"># The layout algorithm that initializes node locations</span></code></pre></div>
<p>This is known as <b><a href ="00--glossary.html#non-standard_evaluation_(nse)_anchor">Non-standard evaluation (NSE)</a></b> in programmer jargon and will be used in many functions throughout this workshop.</p>
</div>
<div id="alpha-diversity" class="section level3">
<h3>Alpha diversity</h3>
<p>Alpha diversity is a measure of the diversity <strong>within</strong> each sample or group of samples. It can be calculated at any rank of the taxonomy, but it is usually calculated at the species or OTU “rank”. There are multiple methods used to calculate a value to represent alpha diversity. The simplest is just the number of species, but the ones used most often factor in how common each species is as well. Below, we calculate the alpha diversity of OTUs using the <b><a href ="00--glossary.html#inverse_simpson_index_anchor">Inverse Simpson Index</a></b> using the package <code>vegan</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(vegan)</code></pre></div>
<pre><code>## Loading required package: permute</code></pre>
<pre><code>## Loading required package: lattice</code></pre>
<pre><code>## This is vegan 2.5-2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hmp_samples<span class="op">$</span>inv_simp &lt;-<span class="st"> </span><span class="kw">diversity</span>(obj<span class="op">$</span>data<span class="op">$</span>tax_data[, hmp_samples<span class="op">$</span>sample_id],
                                  <span class="dt">index =</span> <span class="st">&quot;invsimpson&quot;</span>,
                                  <span class="dt">MARGIN =</span> <span class="dv">2</span>) <span class="co"># What orietation the matrix is in</span></code></pre></div>
<p>Adding this to the sample data table makes it easy to use the sample information in graphing. Lets compare the alpha diversity of samples from males and females using <code>ggplot2</code>, a popular R package for plotting.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)</code></pre></div>
<pre><code>## 
## Attaching package: &#39;ggplot2&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:taxa&#39;:
## 
##     map_data</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(hmp_samples, <span class="kw">aes</span>(<span class="dt">x =</span> sex, <span class="dt">y =</span> inv_simp)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_boxplot</span>()</code></pre></div>
<p><img src="02--quick_example_files/figure-html/unnamed-chunk-18-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Not much difference there, as you might expect. We can also compare body sites:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(hmp_samples, <span class="kw">aes</span>(<span class="dt">x =</span> body_site, <span class="dt">y =</span> inv_simp)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_boxplot</span>()</code></pre></div>
<p><img src="02--quick_example_files/figure-html/unnamed-chunk-19-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>That’s more interesting; skin has much lower diversity than any of the wetter areas, which makes sense. Lets see if that’s a significant difference using <b><a href ="00--glossary.html#analysis_of_variance_(anova)_anchor">analysis of variance (ANOVA)</a></b>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">anova_result &lt;-<span class="st"> </span><span class="kw">aov</span>(inv_simp <span class="op">~</span><span class="st"> </span>body_site, hmp_samples)
<span class="kw">summary</span>(anova_result)</code></pre></div>
<pre class="r-output"><code>##             Df Sum Sq Mean Sq F value Pr(&gt;F)  
## body_site    4   1155  288.63   3.247 0.0201 *
## Residuals   45   4000   88.89                 
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
</code></pre>
<p>That tells that at least one of the body site means is different from the other, but not which one (although we can make a good guess). A <b><a href ="00--glossary.html#tukey's_honest_significant_difference_(hsd)_anchor">Tukey’s Honest Significant Difference (HSD)</a></b> test can compare each site to every other and tell us which are significantly different. Although <b><a href ="00--glossary.html#base_r_anchor">base R</a></b> has a Tukey’s HSD function called <code>TukeyHSD</code>, we will use one from the package <code>agricolae</code> since it supplies grouping codes that are useful for graphing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(agricolae)
tukey_result &lt;-<span class="st"> </span><span class="kw">HSD.test</span>(anova_result, <span class="st">&quot;body_site&quot;</span>, <span class="dt">group =</span> <span class="ot">TRUE</span>)
<span class="kw">print</span>(tukey_result)</code></pre></div>
<pre class="r-output"><code>## $statistics
##    MSerror Df     Mean       CV      MSD
##   88.89491 45 17.94757 52.53306 11.98101
## 
## $parameters
##    test    name.t ntr StudentizedRange alpha
##   Tukey body_site   5         4.018417  0.05
## 
## $means
##         inv_simp       std  r       Min      Max       Q25       Q50       Q75
## Nose   17.662798 10.332621 10  7.281343 37.49351 11.612608 14.592358 16.916579
## Saliva 22.543018  6.230297 10 11.419763 34.20763 19.304662 22.978894 25.294143
## Skin    8.857022  5.585280 10  4.765705 23.98838  5.881698  7.714934  9.019116
## Stool  20.647132 11.392198 10  7.578868 45.00703 11.582727 19.498076 25.109358
## Throat 20.027902 11.743822 10  2.973287 36.94885 11.194047 20.290096 25.959168
## 
## $comparison
## NULL
## 
## $groups
##         inv_simp groups
## Saliva 22.543018      a
## Stool  20.647132     ab
## Throat 20.027902     ab
## Nose   17.662798     ab
## Skin    8.857022      b
## 
## attr(,"class")
## [1] "group"
</code></pre>
<p>We are interested in the <code>$groups</code> table that says which sites are different. With a little tweaking, we can add this data to the graph we made. Lets add some nicer text as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">group_data &lt;-<span class="st"> </span>tukey_result<span class="op">$</span>groups[<span class="kw">order</span>(<span class="kw">rownames</span>(tukey_result<span class="op">$</span>groups)),]
<span class="kw">ggplot</span>(hmp_samples, <span class="kw">aes</span>(<span class="dt">x =</span> body_site, <span class="dt">y =</span> inv_simp)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_text</span>(<span class="dt">data =</span> <span class="kw">data.frame</span>(),
            <span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw">rownames</span>(group_data), <span class="dt">y =</span> <span class="kw">max</span>(hmp_samples<span class="op">$</span>inv_simp) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">label =</span> group_data<span class="op">$</span>groups),
            <span class="dt">col =</span> <span class="st">&#39;black&#39;</span>,
            <span class="dt">size =</span> <span class="dv">10</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_boxplot</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Alpha diversity of human body sites&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;Body site&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;Inverse Simpson Index&quot;</span>)</code></pre></div>
<p><img src="02--quick_example_files/figure-html/unnamed-chunk-22-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>This tells us that samples from the saliva and skin are significantly different from each other, but not significantly different from anything else.</p>
</div>
<div id="comparing-two-treatmentsgroups" class="section level3">
<h3>Comparing two treatments/groups</h3>
<p>Usually we are interested in how groups of samples compare. For example, we might want to know which taxa differ between the nose and throat, or between men and women. The function <code>compare_groups</code> facilitates these comparisons:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>diff_table &lt;-<span class="st"> </span><span class="kw">compare_groups</span>(obj,
                                      <span class="dt">dataset =</span> <span class="st">&quot;tax_abund&quot;</span>,
                                      <span class="dt">cols =</span> hmp_samples<span class="op">$</span>sample_id, <span class="co"># What columns of sample data to use</span>
                                      <span class="dt">groups =</span> hmp_samples<span class="op">$</span>sex) <span class="co"># What category each sample is assigned to</span>
<span class="kw">print</span>(obj<span class="op">$</span>data<span class="op">$</span>diff_table)</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 155 x 7</span><span>
##    taxon_id treatment_1 treatment_2 log2_median_ratio median_diff mean_diff wilcox_p_value
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                   </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>          </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> ab       female      male                    0.         0.       0.             </span><span style='color: #BB0000;'>NaN</span><span>    
## </span><span style='color: #BCBCBC;'> 2</span><span> ac       female      male                    0.380      0.022</span><span style='text-decoration: underline;'>9</span><span>   0.037</span><span style='text-decoration: underline;'>9</span><span>           0.470
## </span><span style='color: #BCBCBC;'> 3</span><span> ad       female      male                   -</span><span style='color: #BB0000;'>0.434</span><span>     -</span><span style='color: #BB0000;'>0.044</span><span style='color: #BB0000;text-decoration: underline;'>9</span><span>  -</span><span style='color: #BB0000;'>0.019</span><span style='color: #BB0000;text-decoration: underline;'>9</span><span>           0.907
## </span><span style='color: #BCBCBC;'> 4</span><span> ae       female      male                   -</span><span style='color: #BB0000;'>1.68</span><span>      -</span><span style='color: #BB0000;'>0.047</span><span style='color: #BB0000;text-decoration: underline;'>5</span><span>  -</span><span style='color: #BB0000;'>0.075</span><span style='color: #BB0000;text-decoration: underline;'>3</span><span>           0.335
## </span><span style='color: #BCBCBC;'> 5</span><span> af       female      male                    0.649      0.116    0.061</span><span style='text-decoration: underline;'>4</span><span>           0.386
## </span><span style='color: #BCBCBC;'> 6</span><span> ag       female      male                    0.         0.      -</span><span style='color: #BB0000;'>0.002</span><span style='color: #BB0000;text-decoration: underline;'>75</span><span>          0.732
## </span><span style='color: #BCBCBC;'> 7</span><span> ah       female      male                    0.         0.      -</span><span style='color: #BB0000;'>0.001</span><span style='color: #BB0000;text-decoration: underline;'>40</span><span>          0.602
## </span><span style='color: #BCBCBC;'> 8</span><span> aj       female      male                    1.24       0.016</span><span style='text-decoration: underline;'>2</span><span>  -</span><span style='color: #BB0000;'>0.012</span><span style='color: #BB0000;text-decoration: underline;'>9</span><span>           0.680
## </span><span style='color: #BCBCBC;'> 9</span><span> ak       female      male                    0.         0.       0.001</span><span style='text-decoration: underline;'>21</span><span>          0.416
## </span><span style='color: #BCBCBC;'>10</span><span> al       female      male                   -</span><span style='color: #BB0000;'>0.542</span><span>     -</span><span style='color: #BB0000;'>0.054</span><span style='color: #BB0000;text-decoration: underline;'>1</span><span>  -</span><span style='color: #BB0000;'>0.021</span><span style='color: #BB0000;text-decoration: underline;'>1</span><span>           0.969
## </span><span style='color: #949494;'># ... with 145 more rows</span><span>
</span></code></pre>
<p>For each taxon, a <b><a href ="00--glossary.html#wilcoxon_rank_sum_test_anchor">Wilcoxon Rank Sum test</a></b> was used to test for differences between the median abundances of samples in each treatment. We can use this information to create what we call a <strong>differential heat tree</strong>, which indicates which taxa are more abundant in each treatment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">999</span>)
<span class="kw">heat_tree</span>(obj, 
          <span class="dt">node_label =</span> taxon_names,
          <span class="dt">node_size =</span> n_obs, <span class="co"># n_obs is a function that calculates, in this case, the number of OTUs per taxon</span>
          <span class="dt">node_color =</span> log2_median_ratio, <span class="co"># A column from `obj$data$diff_table`</span>
          <span class="dt">node_color_interval =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">2</span>), <span class="co"># The range of `log2_median_ratio` to display</span>
          <span class="dt">node_color_range =</span> <span class="kw">c</span>(<span class="st">&quot;cyan&quot;</span>, <span class="st">&quot;gray&quot;</span>, <span class="st">&quot;tan&quot;</span>), <span class="co"># The color palette used</span>
          <span class="dt">node_size_axis_label =</span> <span class="st">&quot;OTU count&quot;</span>,
          <span class="dt">node_color_axis_label =</span> <span class="st">&quot;Log 2 ratio of median proportions&quot;</span>,
          <span class="dt">layout =</span> <span class="st">&quot;davidson-harel&quot;</span>, <span class="co"># The primary layout algorithm</span>
          <span class="dt">initial_layout =</span> <span class="st">&quot;reingold-tilford&quot;</span>) <span class="co"># The layout algorithm that initializes node locations</span></code></pre></div>
<p><img src="02--quick_example_files/figure-html/unnamed-chunk-24-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>In this case, taxa colored tan are more abundant in women and those colored blue are more abundant in men. Note that we have not taken into account statistical significance when showing this, so lets do that. First, we need to <b><a href ="00--glossary.html#multiple_comparison_corrections_anchor">correct for multiple comparisons</a></b>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>diff_table<span class="op">$</span>wilcox_p_value &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(obj<span class="op">$</span>data<span class="op">$</span>diff_table<span class="op">$</span>wilcox_p_value,
                                               <span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>)</code></pre></div>
<p>If we then look at the distribution of p-values, we can see that none are even close to significant:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">range</span>(obj<span class="op">$</span>data<span class="op">$</span>diff_table<span class="op">$</span>wilcox_p_value, <span class="dt">finite =</span> <span class="ot">TRUE</span>) </code></pre></div>
<pre class="r-output"><code>## [1] 0.8029954 1.0000000
</code></pre>
<p>There is no need to graph this, but if there still were some significant differences, we could set any difference that is not significant to zero and repeat the last <code>heat_tree</code> command doing something like:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>diff_table<span class="op">$</span>log2_median_ratio[obj<span class="op">$</span>data<span class="op">$</span>diff_table<span class="op">$</span>wilcox_p_value <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.05</span>] &lt;-<span class="st"> </span><span class="dv">0</span></code></pre></div>
</div>
<div id="comparing-any-number-of-treatmentsgroups" class="section level3">
<h3>Comparing any number of treatments/groups</h3>
<p>A single differential heat tree can compare two treatments (e.g. male and female), but what if you have more? Then we can make a matrix of heat trees, one for each pairwise comparison of treatments like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>diff_table &lt;-<span class="st"> </span><span class="kw">compare_groups</span>(obj, <span class="dt">dataset =</span> <span class="st">&quot;tax_abund&quot;</span>,
                                      <span class="dt">cols =</span> hmp_samples<span class="op">$</span>sample_id, <span class="co"># What columns of sample data to use</span>
                                      <span class="dt">groups =</span> hmp_samples<span class="op">$</span>body_site) <span class="co"># What category each sample is assigned to</span>
<span class="kw">print</span>(obj<span class="op">$</span>data<span class="op">$</span>diff_table)</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 1,550 x 7</span><span>
##    taxon_id treatment_1 treatment_2 log2_median_ratio median_diff mean_diff wilcox_p_value
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                   </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>          </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> ab       Nose        Saliva                   0.        0.        0.        </span><span style='color: #BB0000;'>NaN</span><span>        
## </span><span style='color: #BCBCBC;'> 2</span><span> ac       Nose        Saliva                  -</span><span style='color: #BB0000;'>2.62</span><span>     -</span><span style='color: #BB0000;'>0.167</span><span>    -</span><span style='color: #BB0000;'>0.128</span><span>       0.017</span><span style='text-decoration: underline;'>2</span><span>   
## </span><span style='color: #BCBCBC;'> 3</span><span> ad       Nose        Saliva                  -</span><span style='color: #BB0000;'>7.68</span><span>     -</span><span style='color: #BB0000;'>0.274</span><span>    -</span><span style='color: #BB0000;'>0.265</span><span>       0.000</span><span style='text-decoration: underline;'>163</span><span> 
## </span><span style='color: #BCBCBC;'> 4</span><span> ae       Nose        Saliva                   5.36      0.616     0.595       0.000</span><span style='text-decoration: underline;'>010</span><span>8
## </span><span style='color: #BCBCBC;'> 5</span><span> af       Nose        Saliva                  -</span><span style='color: #BB0000;'>1.23</span><span>     -</span><span style='color: #BB0000;'>0.260</span><span>    -</span><span style='color: #BB0000;'>0.159</span><span>       0.043</span><span style='text-decoration: underline;'>3</span><span>   
## </span><span style='color: #BCBCBC;'> 6</span><span> ag       Nose        Saliva                -</span><span style='color: #BB0000;'>Inf</span><span>        -</span><span style='color: #BB0000;'>0.022</span><span style='color: #BB0000;text-decoration: underline;'>8</span><span>   -</span><span style='color: #BB0000;'>0.043</span><span style='color: #BB0000;text-decoration: underline;'>6</span><span>      0.000</span><span style='text-decoration: underline;'>087</span><span>4
## </span><span style='color: #BCBCBC;'> 7</span><span> ah       Nose        Saliva                   0.        0.        0.        </span><span style='color: #BB0000;'>NaN</span><span>        
## </span><span style='color: #BCBCBC;'> 8</span><span> aj       Nose        Saliva                  -</span><span style='color: #BB0000;'>3.83</span><span>     -</span><span style='color: #BB0000;'>0.103</span><span>    -</span><span style='color: #BB0000;'>0.080</span><span style='color: #BB0000;text-decoration: underline;'>3</span><span>      0.007</span><span style='text-decoration: underline;'>07</span><span>  
## </span><span style='color: #BCBCBC;'> 9</span><span> ak       Nose        Saliva                -</span><span style='color: #BB0000;'>Inf</span><span>        -</span><span style='color: #BB0000;'>0.017</span><span style='color: #BB0000;text-decoration: underline;'>4</span><span>   -</span><span style='color: #BB0000;'>0.017</span><span style='color: #BB0000;text-decoration: underline;'>4</span><span>      0.001</span><span style='text-decoration: underline;'>56</span><span>  
## </span><span style='color: #BCBCBC;'>10</span><span> al       Nose        Saliva                -</span><span style='color: #BB0000;'>Inf</span><span>        -</span><span style='color: #BB0000;'>0.258</span><span>    -</span><span style='color: #BB0000;'>0.248</span><span>       0.000</span><span style='text-decoration: underline;'>149</span><span> 
## </span><span style='color: #949494;'># ... with 1,540 more rows</span><span>
</span></code></pre>
<p>There is a special function to plot this type of data called <code>heat_tree_matrix</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
<span class="kw">heat_tree_matrix</span>(obj,
                 <span class="dt">dataset =</span> <span class="st">&quot;diff_table&quot;</span>,
                 <span class="dt">node_size =</span> n_obs, <span class="co"># n_obs is a function that calculates, in this case, the number of OTUs per taxon</span>
                 <span class="dt">node_label =</span> taxon_names,
                 <span class="dt">node_color =</span> log2_median_ratio, <span class="co"># A column from `obj$data$diff_table`</span>
                 <span class="dt">node_color_range =</span> <span class="kw">diverging_palette</span>(), <span class="co"># The built-in palette for diverging data</span>
                 <span class="dt">node_color_trans =</span> <span class="st">&quot;linear&quot;</span>, <span class="co"># The default is scaled by circle area</span>
                 <span class="dt">node_color_interval =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">3</span>, <span class="dv">3</span>), <span class="co"># The range of `log2_median_ratio` to display</span>
                 <span class="dt">edge_color_interval =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">3</span>, <span class="dv">3</span>), <span class="co"># The range of `log2_median_ratio` to display</span>
                 <span class="dt">node_size_axis_label =</span> <span class="st">&quot;Number of OTUs&quot;</span>,
                 <span class="dt">node_color_axis_label =</span> <span class="st">&quot;Log2 ratio median proportions&quot;</span>,
                 <span class="dt">layout =</span> <span class="st">&quot;davidson-harel&quot;</span>, <span class="co"># The primary layout algorithm</span>
                 <span class="dt">initial_layout =</span> <span class="st">&quot;reingold-tilford&quot;</span>, <span class="co"># The layout algorithm that initializes node locations</span>
                 <span class="dt">output_file =</span> <span class="st">&quot;differential_heat_tree.pdf&quot;</span>) <span class="co"># Saves the plot as a pdf file</span></code></pre></div>
<p><img src="02--quick_example_files/figure-html/unnamed-chunk-29-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>The grey tree on the lower left functions as a key for the unlabeled trees. Each of the smaller trees represent a comparison between body sites in the columns and rows. A taxon colored brown is more abundant in the body site in the column and a taxon colored green is more abundant in body site of the row. For example, Bacteroidetes is more abundant (i.e. has more reads) in the throat samples than the nose samples and this is due to the greater abundance of two genera <em>Prevotella</em> and <em>Porphyromonas</em>. Look at the PDF file saved as “differential_heat_tree.pdf” to explore the details of the plot.</p>
</div>
</div>

<div class="footer_section">
  <div class="footer_content">
    <span xmlns:dct="http://purl.org/dc/terms/" href="http://purl.org/dc/dcmitype/Text" property="dct:title" rel="dct:type">Analysis of Microbiome Community Data in R</span> by <a xmlns:cc="http://creativecommons.org/ns#" href="http://grunwaldlab.cgrb.oregonstate.edu/" property="cc:attributionName" rel="cc:attributionURL">The Grunwald lab and the USDA Horticultural Crops Research Unit</a> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.<br />Based on a work at <a xmlns:dct="http://purl.org/dc/terms/" href="https://github.com/grunwaldlab/analysis_of_microbiome_community_data_in_r" rel="dct:source">https://github.com/grunwaldlab/analysis_of_microbiome_community_data_in_r</a>.
  </div>
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
