<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Manipulating taxonomic data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 61px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 66px;
  margin-top: -66px;
}

.section h2 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h3 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h4 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h5 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h6 {
  padding-top: 66px;
  margin-top: -66px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Analysis of Microbiome Data in R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="02--quick_example.html">Example analysis</a>
    </li>
    <li>
      <a href="03--parsing.html">Getting data into R</a>
    </li>
    <li>
      <a href="04--manipulating.html">Manipulating taxonomic data</a>
    </li>
    <li>
      <a href="05--plotting.html">Plotting taxonomic data</a>
    </li>
    <li>
      <a href="06--quality_control.html">Data quality control</a>
    </li>
    <li>
      <a href="07--diversity_stats.html">Diversity and Ordination</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Appendices
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="introduction_to_r.html">Introduction to R</a>
    </li>
  </ul>
</li>
<li>
  <a href="about.html">
    <span class="fa fa-info"></span>
     
    About
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/analysis_of_microbiome_community_data_in_r">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Manipulating taxonomic data</h1>
<h3 class="subtitle"><em><em>ZSL Foster and NJ Grünwald</em></em></h3>

</div>


<div id="load-example-data" class="section level2">
<h2>Load example data</h2>
<p>If you are starting the workshop at this section, or had problems running code in a previous section, use the following to load the data used in this section. If you have just done the previous section, you can ignore this and proceed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;parsed_data.Rdata&quot;</span>)</code></pre></div>
</div>
<div id="the-challenge-of-hierarchical-data" class="section level2">
<h2>The challenge of hierarchical data</h2>
<p>Tabular data is relatively straight forward to manipulate if you know some R already, but taxonomic data is hierarchical, making it much harder to work with. It gets even more complicated when there are other data mapped to (i.e. associated with) the taxonomy.</p>
<p>For example, if we want to remove a bacterial phylum from the data, do we also remove all of the taxa within that phylum? If so, how do we identify which taxa are in that phylum? Depending on the format the data is in, this might be hard. If other data (e.g. OTU counts) are associated with taxa within that phylum, do we also remove those? After all, even if the phylum and the taxa within it are removed, we still know that all the associated data are bacterial, so do we reassign them to the Bacteria taxon, or just throw those data out? The answer to these kind of questions will depend on our data and what we want to do with it.</p>
</div>
<div id="subsetting-tabular-data" class="section level2">
<h2>Subsetting tabular data</h2>
<p>The functions to subset taxonomic information in <code>taxmap</code> objects (the format our example data is in) provided by the <code>taxa</code> package are modeled after functions in the <code>dplyr</code> package for manipulating tabular data (<span class="citation">@wickham2015dplyr</span>). These <code>dplyr</code> functions are alternatives to the standard R indexing (e.g. <code>[</code>, <code>[[</code>, and <code>$</code>) designed to be be more intuitive and consistent. Since an understanding of these functions will help with the analogous (and more complicated) functions for taxonomic data, we will start by practicing these. Fortunately, we have a tabular data set in need of some subsetting: our sample data.</p>
<p>First, lets take another look at our sample data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(sample_data)</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 1,698 x 16</span><span>
##    SampleID  Name   Plant_ID Type  Experiment Cohort Harvested Age   Site  Treatment Line  Genotype
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>   
## </span><span style='color: #BCBCBC;'> 1</span><span> M1024P17… R_026… R_026    root  fieldBCMA  2008   2011      3     LTM   field     26    ril     
## </span><span style='color: #BCBCBC;'> 2</span><span> M1024P17… R_073… R_073    root  fieldBCMA  2008   2011      3     LTM   field     73    ril     
## </span><span style='color: #BCBCBC;'> 3</span><span> M1024P17… R_088… R_088    root  fieldBCMA  2009   2011      2     LTM   field     88    ril     
## </span><span style='color: #BCBCBC;'> 4</span><span> M1024P18… R_156… R_156    root  fieldBCMA  2009   2011      2     LTM   field     156   ril     
## </span><span style='color: #BCBCBC;'> 5</span><span> M1955P804 1_A_1… 1_A_1    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   par1… PAR     
## </span><span style='color: #BCBCBC;'> 6</span><span> M1956P837 1_A_1… 1_A_12   root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   mah3… MAH     
## </span><span style='color: #BCBCBC;'> 7</span><span> M1957P983 1_A_3… 1_A_3    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   silL… SIL     
## </span><span style='color: #BCBCBC;'> 8</span><span> M1956P845 1_A_4… 1_A_4    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   par1… PAR     
## </span><span style='color: #BCBCBC;'> 9</span><span> M1957P987 1_A_7… 1_A_7    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   par9… PAR     
## </span><span style='color: #BCBCBC;'>10</span><span> M1957P923 1_A_8… 1_A_8    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   mil5… MIL     
## </span><span style='color: #949494;'># ... with 1,688 more rows, and 4 more variables: Block </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, oldPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, newPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>,
## #   Analysis </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
</span></code></pre>
<p>This is a large data set and not all the 1698 samples are used in the main publication of <span class="citation">@wagner2016host</span>. Only the samples in the “ecotypes” experiment are part of the main study. Using standard R indexing we could remove those like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample_data &lt;-<span class="st"> </span>sample_data[sample_data<span class="op">$</span>Experiment <span class="op">==</span><span class="st"> &quot;ecotypes&quot;</span>, ]</code></pre></div>
<p>Using the <code>dplyr</code> functions, the same operation would be:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(dplyr)
sample_data &lt;-<span class="st"> </span><span class="kw">filter</span>(sample_data, Experiment <span class="op">==</span><span class="st"> &quot;ecotypes&quot;</span>)</code></pre></div>
<p>Note how we could use the <code>Experiment</code> column on its own within the <code>filter</code> functions. That is a special property of the <code>filter</code> function and many others in <code>dplyr</code>. If we tried that outside of the function, we would get an error. For example, the following will not work:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">is_ecotype &lt;-<span class="st"> </span>Experiment <span class="op">==</span><span class="st"> &quot;ecotypes&quot;</span></code></pre></div>
<pre><code>## Error in eval(expr, envir, enclos): object &#39;Experiment&#39; not found</code></pre>
<p>Most <code>dplyr</code> functions work this way and so do the analogous functions in <code>taxa</code> for manipulating taxonomic data. If we have multiple filtering criteria, we can simply add more arguments to the functions. In the following code we subset the table to only rows for plants from one of three site and that are 3 years old.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample_data &lt;-<span class="st"> </span><span class="kw">filter</span>(sample_data,
                      Site <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Mah&quot;</span>, <span class="st">&quot;Jam&quot;</span>, <span class="st">&quot;Sil&quot;</span>),
                      Age <span class="op">==</span><span class="st"> </span><span class="dv">3</span>)</code></pre></div>
<p>This is usually equivalent to combining multiple conditions with <code>&amp;</code> in normal subsetting, but is faster for large data sets. It also will not return values for which conditions evaluate to <code>NA</code>, unlike base subsetting.</p>
<p>Up until now, everything we have done using <code>filter</code> can be easily done with base R subsetting as well, but <code>dplyr</code> introduces a few more advanced features that can save time/typing. We wont go into these in detail, but its good to know they exist. One is the concept of “grouping”. When rows are “grouped” by the values in one column, the rows corresponding to each unique value in that column are treated as a unit in some functions. <code>n()</code> is one such function that counts the number of items in each group. We can use this to filter out any plants that don’t have both root and leaf samples by counting the number of times each “Plant_ID” shows up:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample_data &lt;-<span class="st"> </span>sample_data <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(Plant_ID) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">n</span>() <span class="op">==</span><span class="st"> </span><span class="dv">2</span>)</code></pre></div>
<p>This leaves us with 292 samples. Note the use of the <code>%&gt;%</code> “piping” operator. This takes the result of the function that comes before and uses it as the first argument to function that comes after. For example, <code>seq(1:10) %&gt;% length()</code> is the same as <code>length(seq(1:10))</code>. This allows for chaining many commands together in a readable way. Next we will subset the abundance data to just these samples to get the dataset down to a manageable size for a standard personal computer.</p>
</div>
<div id="subsetting-a-taxonomy" class="section level2">
<h2>Subsetting a taxonomy</h2>
<p>The analog to <code>filter</code> for taxonomic data is <code>filter_taxa</code> from the <code>taxa</code> package. This works in a similar way, but has much more options for taxonomic relationships are used and what is done with data mapped to removed taxa. Lets take another look at our data object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   1558 taxa: <span style='color: #00BB00;'>aab</span><span>. Unassigned, </span><span style='color: #00BB00;'>aac</span><span>. Root ... </span><span style='color: #00BB00;'>chx</span><span>. Methanospirillaceae, </span><span style='color: #00BB00;'>chy</span><span>. 
##   1558 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aab</span><span>, </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aac</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aae</span><span> ... </span><span style='color: #00BB00;'>bel</span><span>-&gt;</span><span style='color: #00BB00;'>chw</span><span>, </span><span style='color: #00BB00;'>ays</span><span>-&gt;</span><span style='color: #00BB00;'>chx</span><span>, </span><span style='color: #00BB00;'>bem</span><span>-&gt;</span><span style='color: #00BB00;'>chy</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 47,806 x 1,707</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>aab</span><span>      1              0.       0.       0.       0.       0.       0.       0.
##       2 </span><span style='color: #00BB00;'>ben</span><span>      2             41.      22.       4.     726.  112492.       2.     413.
##       3 </span><span style='color: #00BB00;'>beo</span><span>      3             67.      65.      12.   13514.       1.       4.   13314.
##       # ... with 4.78e+04 more rows, and 1,698 more variables: M1551P90 &lt;dbl&gt;,
##       #   M1551P71 &lt;dbl&gt;, M1551P12 &lt;dbl&gt;, M1551P84 &lt;dbl&gt;, M1551P48 &lt;dbl&gt;, M1551P4 &lt;dbl&gt;,
##       #   M1551P52 &lt;dbl&gt;, M1551P3 &lt;dbl&gt;, M1551P15 &lt;dbl&gt;, M1551P31 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
<p>If we look at the second line, we can see that some taxa do not have names. These are the ones that only had rank information in the classification we parsed in the last section (e.g. <code>o__</code> in <code>&quot;Root;k__Bacteria;p__Chlorobi;c__SJA-28;o__;f__&quot;</code>). Since these do not add any information, lets remove them.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(taxa)
obj &lt;-<span class="st"> </span><span class="kw">filter_taxa</span>(obj, taxon_names <span class="op">!=</span><span class="st"> &quot;&quot;</span>)
<span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   959 taxa: <span style='color: #00BB00;'>aab</span><span>. Unassigned, </span><span style='color: #00BB00;'>aac</span><span>. Root ... </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae, </span><span style='color: #00BB00;'>chx</span><span>. Methanospirillaceae
##   959 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aab</span><span>, </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aac</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aae</span><span> ... </span><span style='color: #00BB00;'>anq</span><span>-&gt;</span><span style='color: #00BB00;'>cho</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>, </span><span style='color: #00BB00;'>ays</span><span>-&gt;</span><span style='color: #00BB00;'>chx</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 47,806 x 1,707</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>aab</span><span>      1              0.       0.       0.       0.       0.       0.       0.
##       2 </span><span style='color: #00BB00;'>ben</span><span>      2             41.      22.       4.     726.  112492.       2.     413.
##       3 </span><span style='color: #00BB00;'>beo</span><span>      3             67.      65.      12.   13514.       1.       4.   13314.
##       # ... with 4.78e+04 more rows, and 1,698 more variables: M1551P90 &lt;dbl&gt;,
##       #   M1551P71 &lt;dbl&gt;, M1551P12 &lt;dbl&gt;, M1551P84 &lt;dbl&gt;, M1551P48 &lt;dbl&gt;, M1551P4 &lt;dbl&gt;,
##       #   M1551P52 &lt;dbl&gt;, M1551P3 &lt;dbl&gt;, M1551P15 &lt;dbl&gt;, M1551P31 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
<p>There were OTUs mapped to the nameless taxa, but no OTUs were filtered out, so what happened to them? By default, they are reassigned to the closest supertaxon that passes the filter. For example, for the classification <code>&quot;Root;k__Bacteria;p__Chlorobi;c__SJA-28;o__;f__&quot;</code>, the OTU would be reassigned to the class “SJA-28”.</p>
<p>Like <code>filter</code>, <code>filter_taxa</code> can use variables contained in the input objects as if they were independent variables. In this case, <code>taxon_names</code> is actually a function that is called when referenced in this way.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(obj<span class="op">$</span><span class="kw">taxon_names</span>())</code></pre></div>
<pre class="r-output"><code>##              aab              aac              aad              aae              aaf 
##     "Unassigned"           "Root"       "Bacteria"        "Archaea" "Proteobacteria" 
##              aag 
## "Actinobacteria"
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">taxon_names</span>(obj))</code></pre></div>
<pre class="r-output"><code>##              aab              aac              aad              aae              aaf 
##     "Unassigned"           "Root"       "Bacteria"        "Archaea" "Proteobacteria" 
##              aag 
## "Actinobacteria"
</code></pre>
<p>Other values that can be used this way include columns in tables and lists/vectors stored in the <code>obj$data</code> list in <code>taxmap</code> objects. Since <code>taxmap</code> objects can store any number of lists, vectors, or tables mapped to a taxonomy, there can be many variables that can be referenced by name. The <code>all_names</code> function returns the name and location of variables that can be used this way:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(obj<span class="op">$</span><span class="kw">all_names</span>(), <span class="dv">20</span>)</code></pre></div>
<pre class="r-output"><code>##                taxon_names                  taxon_ids              taxon_indexes 
##              "taxon_names"                "taxon_ids"            "taxon_indexes" 
##            classifications                n_supertaxa              n_supertaxa_1 
##          "classifications"              "n_supertaxa"            "n_supertaxa_1" 
##                  n_subtaxa                n_subtaxa_1                   n_leaves 
##                "n_subtaxa"              "n_subtaxa_1"                 "n_leaves" 
##                 n_leaves_1                taxon_ranks                    is_root 
##               "n_leaves_1"              "taxon_ranks"                  "is_root" 
##                    is_stem                  is_branch                    is_leaf 
##                  "is_stem"                "is_branch"                  "is_leaf" 
##               is_internode                      n_obs                    n_obs_1 
##             "is_internode"                    "n_obs"                  "n_obs_1" 
##     data$otu_counts$OTU_ID data$otu_counts$M1024P1833 
##                   "OTU_ID"               "M1024P1833"
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(obj<span class="op">$</span><span class="kw">all_names</span>())</code></pre></div>
<pre class="r-output"><code>## [1] 1724
</code></pre>
<p>Next, lets also filter out anything not in Bacteria, since the focus of the study was Bacteria.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj &lt;-<span class="st"> </span><span class="kw">filter_taxa</span>(obj, taxon_names <span class="op">==</span><span class="st"> &quot;Bacteria&quot;</span>, <span class="dt">subtaxa =</span> <span class="ot">TRUE</span>)
<span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   898 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cho</span><span>. wb1_P06, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   898 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>bea</span><span>-&gt;</span><span style='color: #00BB00;'>chi</span><span>, </span><span style='color: #00BB00;'>anq</span><span>-&gt;</span><span style='color: #00BB00;'>cho</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 33,641 x 1,707</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             41.      22.       4.     726.  112492.       2.     413.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3             67.      65.      12.   13514.       1.       4.   13314.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4           3229.   13679.    1832.     951.     113.    2496.     567.
##       # ... with 3.364e+04 more rows, and 1,698 more variables: M1551P90 &lt;dbl&gt;,
##       #   M1551P71 &lt;dbl&gt;, M1551P12 &lt;dbl&gt;, M1551P84 &lt;dbl&gt;, M1551P48 &lt;dbl&gt;, M1551P4 &lt;dbl&gt;,
##       #   M1551P52 &lt;dbl&gt;, M1551P3 &lt;dbl&gt;, M1551P15 &lt;dbl&gt;, M1551P31 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
<p>Note the use of the <code>subtaxa = TRUE</code>; this means that all the subtaxa of taxa that pass the filter should also be preserved. If we left this option off, we would would have only a single taxon with all OTUs assigned to it:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_taxa</span>(obj, taxon_names <span class="op">==</span><span class="st"> &quot;Bacteria&quot;</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   1 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria
##   1 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 33,641 x 1,707</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>aad</span><span>      2             41.      22.       4.     726.  112492.       2.     413.
##       2 </span><span style='color: #00BB00;'>aad</span><span>      3             67.      65.      12.   13514.       1.       4.   13314.
##       3 </span><span style='color: #00BB00;'>aad</span><span>      4           3229.   13679.    1832.     951.     113.    2496.     567.
##       # ... with 3.364e+04 more rows, and 1,698 more variables: M1551P90 &lt;dbl&gt;,
##       #   M1551P71 &lt;dbl&gt;, M1551P12 &lt;dbl&gt;, M1551P84 &lt;dbl&gt;, M1551P48 &lt;dbl&gt;, M1551P4 &lt;dbl&gt;,
##       #   M1551P52 &lt;dbl&gt;, M1551P3 &lt;dbl&gt;, M1551P15 &lt;dbl&gt;, M1551P31 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
<p>Next, lets remove the columns from the abundance matrix that were removed from the sample data at the start of the section.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>otu_counts &lt;-<span class="st"> </span>obj<span class="op">$</span>data<span class="op">$</span>otu_counts[<span class="kw">c</span>(<span class="st">&quot;taxon_id&quot;</span>, <span class="st">&quot;OTU_ID&quot;</span>, sample_data<span class="op">$</span>SampleID)]</code></pre></div>
<p>This leaves us with a data set of 33641 OTUs in 292 samples, classified by 898 taxa:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   898 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cho</span><span>. wb1_P06, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   898 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>bea</span><span>-&gt;</span><span style='color: #00BB00;'>chi</span><span>, </span><span style='color: #00BB00;'>anq</span><span>-&gt;</span><span style='color: #00BB00;'>cho</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 33,641 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 3.364e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<div id="subsetting-data-classified-by-a-taxonomy" class="section level2">
<h2>Subsetting data classified by a taxonomy</h2>
<p>In the previous section we subset the taxonomy in a <code>taxmap</code> object using <code>filter_taxa</code>. By default, when taxa are removed, the OTUs assigned to them were reassigned to a supertaxon or removed as well. We can also subset the OTU table directly and remove any taxa that are no longer needed using the funtion <code>filter_obs</code>. The “obs” in the function name stands for “observations”. Unlike <code>filter</code>, which you used previously to subset tables, <code>filter_obs</code> works on a data set in a <code>taxmap</code> object, like our OTU table.</p>
<p>In the last section we removed samples (i.e. columns in the OTU table) that did not appear in the <span class="citation">@wagner2016host</span> publication. This probably means that some of the OTUs only existed in those samples and now have no reads in the remaining samples. Lets remove those to make the data set more managable. First we need to find which OTUs (i.e. rows) no have no reads. We can do that by using the function <code>rowSums</code> to get the number of reads in all samples for each OTU.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">has_no_reads &lt;-<span class="st"> </span><span class="kw">rowSums</span>(obj<span class="op">$</span>data<span class="op">$</span>otu_counts[, sample_data<span class="op">$</span>SampleID]) <span class="op">==</span><span class="st"> </span><span class="dv">0</span></code></pre></div>
<p>Since the table <code>otu_counts</code> also has the taxon ID and OTU ID columns we need to subset the columns to just the samples before using <code>rowSums</code>. The command above results in a <code>logical</code> vector, which is a series of <code>TRUE</code>/<code>FALSE</code> values that can be used to subset things. We can use <code>sum</code> to count the number of <code>TRUE</code> values and hence the number of OTUs without reads.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(has_no_reads)</code></pre></div>
<pre class="r-output"><code>## [1] 11957
</code></pre>
<p>There are 11957 of 33641 OTUs that now have no reads. Now that we know which they are, we can remove them using <code>filter_obs</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, <span class="op">!</span><span class="st"> </span>has_no_reads)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   898 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cho</span><span>. wb1_P06, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   898 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>bea</span><span>-&gt;</span><span style='color: #00BB00;'>chi</span><span>, </span><span style='color: #00BB00;'>anq</span><span>-&gt;</span><span style='color: #00BB00;'>cho</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 2.168e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
<p>Note how this removed rows in the “otu_counts” table but did not remove any taxa. To remove taxa that are no longer observed in any OTUs, we need to add the <code>drop_taxa</code> option:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj &lt;-<span class="st"> </span><span class="kw">filter_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, <span class="op">!</span><span class="st"> </span>has_no_reads, <span class="dt">drop_taxa =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>If there were other tables in <code>obj$data</code> the data for those taxa would have been removed as well unless <code>drop_obs = FALSE</code> was added. Since there is only the one table, we dont need to worry about that.</p>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<div id="subsetting-tabular-data-1" class="section level3">
<h3>Subsetting tabular data</h3>
<p>In these exercises, we will be using the table <code>sample_data</code> from the analysis above.</p>
<p><strong>1)</strong> Try to subset <code>sample_data</code> to just “leaf” samples using standard R indexing (e.g. <code>my_table[rows, columns]</code>).</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-21">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-21" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample_data[sample_data<span class="op">$</span>Type <span class="op">==</span><span class="st"> &quot;leaf&quot;</span>, ]</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 146 x 16</span><span>
## </span><span style='color: #949494;'># Groups:   Plant_ID [146]</span><span>
##    SampleID  Name   Plant_ID Type  Experiment Cohort Harvested Age   Site  Treatment Line  Genotype
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>   
## </span><span style='color: #BCBCBC;'> 1</span><span> M1981P563 8_005… 8_0053   leaf  ecotypes   2008   2011      3     Mah   field     MAH.… MAH     
## </span><span style='color: #BCBCBC;'> 2</span><span> M1980P502 8_007… 8_0074   leaf  ecotypes   2008   2011      3     Mah   field     JAM.… JAM     
## </span><span style='color: #BCBCBC;'> 3</span><span> M1981P599 8_008… 8_0087   leaf  ecotypes   2008   2011      3     Mah   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'> 4</span><span> M1981P587 8_009… 8_0090   leaf  ecotypes   2008   2011      3     Mah   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'> 5</span><span> M1979P447 8_009… 8_0096   leaf  ecotypes   2008   2011      3     Mah   field     PAR.… PAR     
## </span><span style='color: #BCBCBC;'> 6</span><span> M1982P661 8_014… 8_0149   leaf  ecotypes   2008   2011      3     Jam   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'> 7</span><span> M1980P462 8_015… 8_0157   leaf  ecotypes   2008   2011      3     Jam   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'> 8</span><span> M1980P471 8_015… 8_0158   leaf  ecotypes   2008   2011      3     Jam   field     JAM.… JAM     
## </span><span style='color: #BCBCBC;'> 9</span><span> M1982P655 8_016… 8_0167   leaf  ecotypes   2008   2011      3     Jam   field     PAR.… PAR     
## </span><span style='color: #BCBCBC;'>10</span><span> M1981P623 8_017… 8_0174   leaf  ecotypes   2008   2011      3     Jam   field     MAH.… MAH     
## </span><span style='color: #949494;'># ... with 136 more rows, and 4 more variables: Block </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, oldPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, newPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>,
## #   Analysis </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
</span></code></pre>
</div>
<p><br /></p>
<p><strong>2)</strong> Try to do the same thing using the <code>filter</code> function from the <code>dplyr</code> package.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-22">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-22" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter</span>(sample_data, Type <span class="op">==</span><span class="st"> &quot;leaf&quot;</span>)</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 146 x 16</span><span>
## </span><span style='color: #949494;'># Groups:   Plant_ID [146]</span><span>
##    SampleID  Name   Plant_ID Type  Experiment Cohort Harvested Age   Site  Treatment Line  Genotype
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>   
## </span><span style='color: #BCBCBC;'> 1</span><span> M1981P563 8_005… 8_0053   leaf  ecotypes   2008   2011      3     Mah   field     MAH.… MAH     
## </span><span style='color: #BCBCBC;'> 2</span><span> M1980P502 8_007… 8_0074   leaf  ecotypes   2008   2011      3     Mah   field     JAM.… JAM     
## </span><span style='color: #BCBCBC;'> 3</span><span> M1981P599 8_008… 8_0087   leaf  ecotypes   2008   2011      3     Mah   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'> 4</span><span> M1981P587 8_009… 8_0090   leaf  ecotypes   2008   2011      3     Mah   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'> 5</span><span> M1979P447 8_009… 8_0096   leaf  ecotypes   2008   2011      3     Mah   field     PAR.… PAR     
## </span><span style='color: #BCBCBC;'> 6</span><span> M1982P661 8_014… 8_0149   leaf  ecotypes   2008   2011      3     Jam   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'> 7</span><span> M1980P462 8_015… 8_0157   leaf  ecotypes   2008   2011      3     Jam   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'> 8</span><span> M1980P471 8_015… 8_0158   leaf  ecotypes   2008   2011      3     Jam   field     JAM.… JAM     
## </span><span style='color: #BCBCBC;'> 9</span><span> M1982P655 8_016… 8_0167   leaf  ecotypes   2008   2011      3     Jam   field     PAR.… PAR     
## </span><span style='color: #BCBCBC;'>10</span><span> M1981P623 8_017… 8_0174   leaf  ecotypes   2008   2011      3     Jam   field     MAH.… MAH     
## </span><span style='color: #949494;'># ... with 136 more rows, and 4 more variables: Block </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, oldPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, newPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>,
## #   Analysis </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
</span></code></pre>
</div>
<p><br /></p>
<p><strong>3)</strong> Using a single call to the <code>filter</code> function, get all leaf samples from site “Jam” that are of genotype “SIL” or “MIL”.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-23">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-23" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter</span>(sample_data,
       Type <span class="op">==</span><span class="st"> &quot;leaf&quot;</span>,
       Site <span class="op">==</span><span class="st"> &quot;Jam&quot;</span>,
       Genotype <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;SIL&quot;</span>, <span class="st">&quot;MIL&quot;</span>))</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 14 x 16</span><span>
## </span><span style='color: #949494;'># Groups:   Plant_ID [14]</span><span>
##    SampleID  Name   Plant_ID Type  Experiment Cohort Harvested Age   Site  Treatment Line  Genotype
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>   
## </span><span style='color: #BCBCBC;'> 1</span><span> M1982P661 8_014… 8_0149   leaf  ecotypes   2008   2011      3     Jam   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'> 2</span><span> M1980P462 8_015… 8_0157   leaf  ecotypes   2008   2011      3     Jam   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'> 3</span><span> M1982P711 8_032… 8_0326   leaf  ecotypes   2008   2011      3     Jam   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'> 4</span><span> M1979P439 8_060… 8_0609   leaf  ecotypes   2008   2011      3     Jam   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'> 5</span><span> M1981P586 8_061… 8_0612   leaf  ecotypes   2008   2011      3     Jam   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'> 6</span><span> M1980P481 8_077… 8_0775   leaf  ecotypes   2008   2011      3     Jam   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'> 7</span><span> M1982P664 8_080… 8_0803   leaf  ecotypes   2008   2011      3     Jam   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'> 8</span><span> M1981P582 8_102… 8_1023   leaf  ecotypes   2008   2011      3     Jam   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'> 9</span><span> M1982P663 8_104… 8_1048   leaf  ecotypes   2008   2011      3     Jam   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'>10</span><span> M1981P639 8_138… 8_1382   leaf  ecotypes   2008   2011      3     Jam   field     MIL.… MIL     
## </span><span style='color: #BCBCBC;'>11</span><span> M1981P626 8_138… 8_1388   leaf  ecotypes   2008   2011      3     Jam   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'>12</span><span> M1981P622 9_017… 9_0171   leaf  ecotypes   2009   2012      3     Jam   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'>13</span><span> M1982P718 9_079… 9_0799   leaf  ecotypes   2009   2012      3     Jam   field     SILL… SIL     
## </span><span style='color: #BCBCBC;'>14</span><span> M1554P181 9_136… 9_1364   leaf  ecotypes   2009   2012      3     Jam   field     MIL.… MIL     
## </span><span style='color: #949494;'># ... with 4 more variables: Block </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, oldPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, newPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, Analysis </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
</span></code></pre>
</div>
<p><br /></p>
</div>
<div id="subsetting-a-taxonomy-1" class="section level3">
<h3>Subsetting a taxonomy</h3>
<p>In these exercises, we will be using the <code>taxmap</code> object <code>obj</code> from the analysis above. For the questions below, do not overwrite the <code>obj</code> with result of filtering (i.e. don’t assign the result to <code>obj</code>).</p>
<p><strong>1)</strong> The function <code>n_obs</code> counts the number of items in the first dataset (the OTU table in our case) assigned to each taxon. Try running <code>n_obs(obj)</code> to see how many OTUs are within each taxon.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-24">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-24" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">n_obs</span>(obj)</code></pre></div>
<pre class="r-output"><code>##   aad   aaf   aag   aah   aai   aaj   aak   aal   aam   aan   aao   aap   aaq   aar   aas   aau 
## 21684  7771  1781   443  1721   501   867  1150   417  1110  2367   106   382    68   719    20 
##   aav   aaw   aax   aay   aaz   aba   abb   abc   abd   abe   abf   abg   abi   abj   abk   abl 
##     8    17   426   448   158    66    42   163    23    11    32    61   444    86    56     6 
##   abm   abn   abo   abq   abr   abs   abt   abw   abx   aby   abz   acc   acg   aci   acr   act 
##     9    86     2    47     1    10    12     1     6     1     1     4     4     1     1     1 
##   acu   acx   add   ade   adf   adg   adh   adi   adj   adk   adl   adm   adn   ado   adp   adq 
##     1     3  2383   706   233  2015   327   535  2612   430   677    17   349   573   308   188 
##   adr   ads   adt   adu   adv   adw   adx   ady   adz   aea   aeb   aec   aed   aee   aeg   aeh 
##   287    44    20   292   149   177    84   382   227  1666   149   135   254   208   207    92 
##   aei   aej   aek   ael   aem   aen   aeo   aep   aeq   aer   aes   aet   aev   aex   aey   aez 
##    56     6    78    67   324    26    94    83    77   719   105     9    14    56    32   504 
##   afa   afb   afc   afe   aff   afg   afh   afi   afj   afk   afl   afm   afo   afp   afq   afr 
##    19   112    33     5   384   137     8    66     9   158    12    17    37     9    12   150 
##   afs   aft   afu   afv   afw   afx   afy   aga   agb   agc   agd   age   agg   agi   agk   agl 
##    95    30   193    54    56    21    23    81    32    61     1    11    55    14   286    16 
##   agm   agn   ago   agp   agq   agr   ags   agt   agu   agw   agx   agy   agz   ahb   ahc   ahd 
##    41   227     6    39    65    82     5    56    24     3     8     8     4    15    21     9 
##   ahe   ahf   ahg   ahh   ahk   ahn   aho   ahp   ahr   ahs   aht   ahu   ahw   ahx   ahy   aib 
##    14    72     2     3     1    17    11    13     9    30    40     4    16     2    48     7 
##   aie   aif   aig   aih   aii   aij   aim   ain   aip   aiq   air   aiu   aiw   aix   aiy   aja 
##     8    20     8     1     7     6     2    15     5     5    13     2     3     5    20     6 
##   ajb   ajc   ajg   ajm   ajn   ajo   ajq   ajw   ajz   akb   akg   akl   akw   akx   ame   amg 
##     1     2    14     4     1     4     2     3     2     4     1     1     1     1     1   482 
##   amh   ami   amj   amk   aml   amm   amn   amo   amp   amq   amr   ams   amt   amu   amv   amw 
##   181   482   694    78    44   327   232  1205   430   677   561    25   267    23    18    54 
##   amx   amy   amz   ana   anb   anc   and   ane   anf   ang   ani   anj   ank   anl   anm   ann 
##    11    86    27   310   345    75   308   188   168    52    19    97    16   177    71    64 
##   ano   anq   anr   ans   ant   anu   anv   anw   anx   any   aoa   aod   aoe   aof   aog   aoh 
##  1467   382   194   250   139    30    33   252   137   227     8    33    69    26    10    56 
##   aoi   aoj   aok   aom   aon   aoo   aoq   aor   aos   aot   aou   aov   aow   aox   aoy   apa 
##     6  1284    30     6    26   323    94   399    59    83    57    42   711    44     9    14 
##   apc   apd   ape   apg   aph   api   apj   apk   apl   apm   apn   app   apq   apr   aps   apt 
##    37    32   248   109     3   283    67    37    48   369    19    69    45    33    17    25 
##   apv   apw   apx   apy   aqa   aqb   aqf   aqh   aqi   aqj   aqk   aql   aqm   aqn   aqo   aqr 
##    23    23    20    40     5     8    24    20     5   153    13    67    18   104    12    78 
##   aqt   aqu   aqx   aqy   aqz   ara   arb   ard   are   arf   arg   arh   ari   arj   ark   arm 
##    12    34    23    37    31    78    78    20     4    46   141    54     9     3    12    40 
##   aro   arp   arq   ars   arw   arx   asa   asb   ase   asf   asg   asj   asl   asn   asp   asr 
##     2   147    21    47     3    32    15    18    11    33    14    25    42     9    14     9 
##   ast   asu   asy   atc   atd   ate   atf   atg   atj   atl   atn   atp   atq   atr   att   atv 
##     5    26     9     7    54    19     2     7     3    10     8     2     1    16    31     5 
##   atx   aty   aua   aue   auf   aug   auh   aui   auj   aul   auo   auq   aus   aut   auu   auw 
##     9    52     3    52     2     9     3     1     7    13     2     1     2     3    11     5 
##   aux   ava   avb   avc   ave   avi   avl   avm   avo   avq   avs   avv   avx   avz   awb   awe 
##     1     4    30    61     9     4     5     8     5     5     8     2     1     9    11     6 
##   awi   awj   awk   awn   awq   awr   awt   awu   awy   axa   axg   axj   axo   axq   axr   axt 
##     3     4     1     2    16     5     1     1     2     2     2     2     4     2     1     4 
##   axv   axx   axz   ayf   ayg   ayi   ayl   ayq   ayr   ayw   ayx   aze   azg   azh   azi   azp 
##     2     6     1     1     5     2     1     4     1     4     2     2     2     1     2     2 
##   azq   azv   azw   azx   azz   bam   baq   bay   bba   bby   bde   bds   bef   ben   beo   bep 
##     1     2     6     1     1     1     1     1     1     1     1     1     1   156   159    42 
##   beq   bes   bet   beu   bev   bew   bex   bey   bfa   bfb   bfc   bfd   bfe   bff   bfg   bfh 
##    13    44    27   103    47    39    14    97    63   373   614    12   307    43    17    25 
##   bfi   bfj   bfk   bfl   bfn   bfp   bfq   bfr   bfs   bft   bfu   bfv   bfw   bfx   bfy   bfz 
##   147   101    23    13    22   112    13   120     4    11    86    66   171    16    29    35 
##   bgc   bge   bgf   bgg   bgj   bgk   bgl   bgm   bgn   bgo   bgq   bgr   bgs   bgt   bgx   bgz 
##   305   113    33   117    11    16    44    39   227    46    88    16    47    64    66    23 
##   bha   bhb   bhc   bhe   bhf   bhg   bhj   bhm   bhn   bho   bhp   bhq   bhs   bhx   bhz   bia 
##   250    38    36     9    30   120    11    75    19   143    89    34    35    14    36     2 
##   bib   bid   bie   big   bih   bij   bik   bil   bio   bip   bis   bit   biu   biv   biw   bix 
##     8    45    89    10   143     6    14   990     1     7    62    35    16    67    10    90 
##   biy   biz   bja   bjb   bjc   bjd   bje   bjf   bjg   bjh   bji   bjk   bjl   bjm   bjn   bjp 
##    75    59    18    83   847    16    17     2    57    10   171     2   494    33     9    14 
##   bjv   bjy   bkd   bke   bkh   bki   bkl   bkn   bko   bkp   bkt   bku   bkv   bky   bkz   blb 
##   109    41    56    11    49    45    12    92    13     3    23    19   324     5     5    39 
##   blc   bld   blf   bli   bll   blm   bln   blp   blq   blr   bls   blu   blv   blw   blx   bma 
##     3   115     6     7    13    43   293    19     5   153    15    59     5    19    26    12 
##   bmc   bmh   bmk   bml   bmp   bmq   bmv   bnb   bnf   bni   bnl   bnn   bnp   bnt   bnx   bnz 
##     2    29   315     4     7    30     3     2    47     2    38    19    17     4     3    16 
##   boa   boc   boe   boh   boj   bom   bon   boo   bor   bos   bou   boy   bpa   bpb   bpf   bpi 
##     2    42     9    12     3    11     3    33     3     7     8    19     6    12     7     6 
##   bpj   bpk   bpq   bpv   bpw   bpy   bqc   bqj   bqk   bqm   bqp   bqq   bqr   bqs   bqt   bqw 
##     5     1     9    48    10     7     6     5     1     2    11    26    13    20    36     1 
##   bqy   bqz   brb   brd   brh   brm   brn   bro   brq   brr   brs   brw   bsb   bsc   bsd   bsf 
##     2     5     3     4     1    52     1     8     3     4     1     1     5     1     3     7 
##   bsj   bsl   bsq   bsr   bst   bsw   bsy   bsz   bta   btb   bte   btg   bti   btk   btr   btu 
##     1     1     3    10     2     7     1     2    12     8     1     1     4     3     9     1 
##   btw   btx   bua   buf   bug   buh   bui   buk   bul   buo   buq   bux   buz   bvc   bvd   bvf 
##     3     2     7     2     5     2     1     2     9     1     2     5     2     7     4     3 
##   bvg   bvh   bvi   bvs   bvu   bvv   bvw   bvz   bwd   bwe   bwf   bwh   bwk   bwn   bwq   bwr 
##    19     1     1     1     1     8     1     7     6     3     2     2     2     8     2     2 
##   bwu   bww   bxc   bxe   bxp   bxs   bxw   bxz   bye   byf   byh   byi   byk   byl   bym   byr 
##     1     2     5     1     4     1     3     3     1     1     1     5     2     1     3     1 
##   byt   byy   bza   bzb   bze   bzh   bzi   bzl   bzm   bzt   cab   cag   cal   cao   car   cau 
##     2     3     3     3     2     1     5     4     1     1     1     1     1     1     1     1 
##   caw   cbb   cbc   cbr   cbx   cbz   ccd   ccw   ccz   cds   ceq   cfe   cfp   cgf   cgz   chs 
##     2     1     3     1     1     1     1     1     2     1     1     1     3     1     1     1
</code></pre>
</div>
<p><br /></p>
<p><strong>2)</strong> Like <code>taxon_names</code>, <code>n_obs</code> can be used within calls to <code>filter_taxa</code> as if it were a variable. Try to use <code>filter_taxa</code> to remove all taxa that have less than 100 OTUs.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-25">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-25" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_taxa</span>(obj, n_obs <span class="op">&gt;=</span><span class="st"> </span><span class="dv">100</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   122 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>bmk</span><span>. Legionellaceae
##   122 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>aoj</span><span>-&gt;</span><span style='color: #00BB00;'>bln</span><span>, </span><span style='color: #00BB00;'>aqj</span><span>-&gt;</span><span style='color: #00BB00;'>blr</span><span>, </span><span style='color: #00BB00;'>ano</span><span>-&gt;</span><span style='color: #00BB00;'>bmk</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 2.168e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>3a)</strong> Subset <code>obj</code> to just “Actinobacteria” and <em>all of its subtaxa</em>.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-26">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-26" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_taxa</span>(obj, taxon_names <span class="op">==</span><span class="st"> &quot;Actinobacteria&quot;</span>, <span class="dt">subtaxa =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   72 taxa: <span style='color: #00BB00;'>aag</span><span>. Actinobacteria, </span><span style='color: #00BB00;'>ade</span><span>. Actinobacteria ... </span><span style='color: #00BB00;'>cfp</span><span>. Dermacoccaceae
##   72 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aag</span><span>-&gt;</span><span style='color: #00BB00;'>ade</span><span>, </span><span style='color: #00BB00;'>aag</span><span>-&gt;</span><span style='color: #00BB00;'>ado</span><span>, </span><span style='color: #00BB00;'>aag</span><span>-&gt;</span><span style='color: #00BB00;'>ady</span><span> ... </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>cbz</span><span>, </span><span style='color: #00BB00;'>anq</span><span>-&gt;</span><span style='color: #00BB00;'>cfe</span><span>, </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>cfp</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 1,781 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>beq</span><span>      6            70.      4055.      206.     2415.        2.     2901.
##       2 </span><span style='color: #00BB00;'>bev</span><span>      11           45.      1994.       17.     1741.        0.     1598.
##       3 </span><span style='color: #00BB00;'>bew</span><span>      12          711.      3553.      501.     2220.        1.     2410.
##       # ... with 1,778 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>3b)</strong> Why are there fewer OTUs in the result of the above filtering?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-27">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-27" class="collapse">
Because the OTUs assinged to taxa besides <em>Actinobacteria</em> or one of its subtaxa were removed.
</div>
<p><br /></p>
<p><strong>3c)</strong> Look at the documentaion for <code>filter_taxa</code> by typing <code>?filter_taxa</code>. See if you can modify the previous command so that no OTUs are removed from the table, but the taxonomy is still subset to “Actinobacteria” and all of its subtaxa.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-28">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-28" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_taxa</span>(obj, taxon_names <span class="op">==</span><span class="st"> &quot;Actinobacteria&quot;</span>, <span class="dt">subtaxa =</span> <span class="ot">TRUE</span>, <span class="dt">drop_obs =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   72 taxa: <span style='color: #00BB00;'>aag</span><span>. Actinobacteria, </span><span style='color: #00BB00;'>ade</span><span>. Actinobacteria ... </span><span style='color: #00BB00;'>cfp</span><span>. Dermacoccaceae
##   72 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aag</span><span>-&gt;</span><span style='color: #00BB00;'>ade</span><span>, </span><span style='color: #00BB00;'>aag</span><span>-&gt;</span><span style='color: #00BB00;'>ado</span><span>, </span><span style='color: #00BB00;'>aag</span><span>-&gt;</span><span style='color: #00BB00;'>ady</span><span> ... </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>cbz</span><span>, </span><span style='color: #00BB00;'>anq</span><span>-&gt;</span><span style='color: #00BB00;'>cfe</span><span>, </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>cfp</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>NA</span><span>       2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>NA</span><span>       3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>NA</span><span>       4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 2.168e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>4)</strong> The function <code>taxon_ranks</code> returns the ranks for each taxon and it can be used like <code>taxon_names</code> in calls to <code>filter_taxa</code>. Look at the result of <code>taxon_ranks(obj)</code>. Subset <code>obj</code> to phyla (“p” in our data) <em>and their supertaxa</em>.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-29">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-29" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_taxa</span>(obj, taxon_ranks <span class="op">==</span><span class="st"> &quot;p&quot;</span>, <span class="dt">supertaxa =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   50 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>acu</span><span>. GAL15, </span><span style='color: #00BB00;'>acx</span><span>. Kazan-3B-28
##   50 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>act</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>acu</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>acx</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>aaf</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>aaf</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>aaf</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 2.168e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>5a)</strong> Try to remove “Actinobacteria” and all of its subtaxa. You will probably need to look at the documentation for <code>filter_taxa</code> to figure out how to do this.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-30">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-30" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_taxa</span>(obj, taxon_names <span class="op">==</span><span class="st"> &quot;Actinobacteria&quot;</span>, <span class="dt">subtaxa =</span> <span class="ot">TRUE</span>, <span class="dt">invert =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   632 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cgz</span><span>. S24-7, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   632 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aai</span><span> ... </span><span style='color: #00BB00;'>amg</span><span>-&gt;</span><span style='color: #00BB00;'>cgf</span><span>, </span><span style='color: #00BB00;'>arh</span><span>-&gt;</span><span style='color: #00BB00;'>cgz</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 2.168e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>5b)</strong> Why were no OTUs removed when “Actinobacteria” and all of its subtaxa were removed?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-31">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-31" class="collapse">
Because the OTUs that were assigned to “Actinobacteria” were reassigned to “Bacteria”, its supertaxon.
</div>
<p><br /></p>
<p><strong>5c)</strong> Try to modify the previous command to remove the OTUs that are assigned to “Actinobacteria”.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-32">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-32" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_taxa</span>(obj, taxon_names <span class="op">==</span><span class="st"> &quot;Actinobacteria&quot;</span>,
            <span class="dt">subtaxa =</span> <span class="ot">TRUE</span>,
            <span class="dt">invert =</span> <span class="ot">TRUE</span>,
            <span class="dt">reassign_obs =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   632 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cgz</span><span>. S24-7, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   632 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aai</span><span> ... </span><span style='color: #00BB00;'>amg</span><span>-&gt;</span><span style='color: #00BB00;'>cgf</span><span>, </span><span style='color: #00BB00;'>arh</span><span>-&gt;</span><span style='color: #00BB00;'>cgz</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 19,903 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 1.99e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
</div>
<div id="subsetting-data-assocaited-with-a-taxonomy" class="section level3">
<h3>Subsetting data assocaited with a taxonomy</h3>
<p>In these exercises, we will be using the <code>taxmap</code> object <code>obj</code> from the analysis above. For the questions below, do not overwrite the <code>obj</code> with result of filtering (i.e. don’t assign the result to <code>obj</code>).</p>
<p><strong>1a)</strong> Use <code>filter_obs</code> to remove all rows of “otu_counts” that have less than 5 reads in the first sample, “M1981P563” without removing any taxa in the taxonomy.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-33">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-33" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, M1981P563 <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   704 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cgz</span><span>. S24-7, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   704 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>amg</span><span>-&gt;</span><span style='color: #00BB00;'>cgf</span><span>, </span><span style='color: #00BB00;'>arh</span><span>-&gt;</span><span style='color: #00BB00;'>cgz</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 235 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 232 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;, M1982P729 &lt;dbl&gt;,
##       #   M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;, M1978P293 &lt;dbl&gt;,
##       #   M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;, M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Also works: filter_obs(obj, &quot;otu_counts&quot;, obj$data$otu_counts$M1981P563 &gt; 5)</span></code></pre></div>
</div>
<p><br /></p>
<p><strong>1b)</strong> Modify the previous command to remove taxa as well as rows in the table.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-34">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-34" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, M1981P563 <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>, <span class="dt">drop_taxa =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   101 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>btg</span><span>. Bogoriellaceae
##   101 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>aor</span><span>-&gt;</span><span style='color: #00BB00;'>bkv</span><span>, </span><span style='color: #00BB00;'>aoj</span><span>-&gt;</span><span style='color: #00BB00;'>bln</span><span>, </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>btg</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 235 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 232 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;, M1982P729 &lt;dbl&gt;,
##       #   M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;, M1978P293 &lt;dbl&gt;,
##       #   M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;, M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>2a)</strong> The function <code>calc_prop_samples</code> from the <code>metacoder</code> package returns the proportion of samples each OTU appears in (i.e. has a non-zero count):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(metacoder)
<span class="kw">calc_prop_samples</span>(obj, <span class="dt">dataset =</span> <span class="st">&quot;otu_counts&quot;</span>, <span class="dt">groups =</span> sample_data<span class="op">$</span>Type)</code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    M1981P563, M1977P1709, M1980P502, M1981P606 ... M1955P747, M1978P342, M1955P778</code></pre>
<pre><code>## Calculating the proportion of samples with a value greater than 0 for 292 columns in 2 groups for 21684 observations</code></pre>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 21,684 x 3</span><span>
##    taxon_id  leaf  root
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> ben      0.863 0.815
## </span><span style='color: #BCBCBC;'> 2</span><span> beo      1.00  1.00 
## </span><span style='color: #BCBCBC;'> 3</span><span> ben      1.00  1.00 
## </span><span style='color: #BCBCBC;'> 4</span><span> bep      0.932 1.00 
## </span><span style='color: #BCBCBC;'> 5</span><span> beq      0.911 1.00 
## </span><span style='color: #BCBCBC;'> 6</span><span> amk      1.00  1.00 
## </span><span style='color: #BCBCBC;'> 7</span><span> bes      0.911 1.00 
## </span><span style='color: #BCBCBC;'> 8</span><span> bet      0.979 0.986
## </span><span style='color: #BCBCBC;'> 9</span><span> beu      0.979 0.993
## </span><span style='color: #BCBCBC;'>10</span><span> bev      0.925 1.00 
## </span><span style='color: #949494;'># ... with 21,674 more rows</span><span>
</span></code></pre>
<p>Using the <code>groups</code> option, it can do the calculation on columns grouped by treatment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">calc_prop_samples</span>(obj, <span class="dt">dataset =</span> <span class="st">&quot;otu_counts&quot;</span>, <span class="dt">groups =</span> sample_data<span class="op">$</span>Type)</code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    M1981P563, M1977P1709, M1980P502, M1981P606 ... M1955P747, M1978P342, M1955P778</code></pre>
<pre><code>## Calculating the proportion of samples with a value greater than 0 for 292 columns in 2 groups for 21684 observations</code></pre>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 21,684 x 3</span><span>
##    taxon_id  leaf  root
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> ben      0.863 0.815
## </span><span style='color: #BCBCBC;'> 2</span><span> beo      1.00  1.00 
## </span><span style='color: #BCBCBC;'> 3</span><span> ben      1.00  1.00 
## </span><span style='color: #BCBCBC;'> 4</span><span> bep      0.932 1.00 
## </span><span style='color: #BCBCBC;'> 5</span><span> beq      0.911 1.00 
## </span><span style='color: #BCBCBC;'> 6</span><span> amk      1.00  1.00 
## </span><span style='color: #BCBCBC;'> 7</span><span> bes      0.911 1.00 
## </span><span style='color: #BCBCBC;'> 8</span><span> bet      0.979 0.986
## </span><span style='color: #BCBCBC;'> 9</span><span> beu      0.979 0.993
## </span><span style='color: #BCBCBC;'>10</span><span> bev      0.925 1.00 
## </span><span style='color: #949494;'># ... with 21,674 more rows</span><span>
</span></code></pre>
<p>Run the following code to add this as a table in your <code>taxmap</code> object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>sample_props &lt;-<span class="st"> </span><span class="kw">calc_prop_samples</span>(obj, <span class="dt">dataset =</span> <span class="st">&quot;otu_counts&quot;</span>, <span class="dt">groups =</span> sample_data<span class="op">$</span>Type)</code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    M1981P563, M1977P1709, M1980P502, M1981P606 ... M1955P747, M1978P342, M1955P778</code></pre>
<pre><code>## Calculating the proportion of samples with a value greater than 0 for 292 columns in 2 groups for 21684 observations</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   704 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cgz</span><span>. S24-7, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   704 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>amg</span><span>-&gt;</span><span style='color: #00BB00;'>cgf</span><span>, </span><span style='color: #00BB00;'>arh</span><span>-&gt;</span><span style='color: #00BB00;'>cgz</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 2.168e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>sample_props</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 3</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span> leaf  root
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      0.863 0.815
##       2 </span><span style='color: #00BB00;'>beo</span><span>      1.00  1.00 
##       3 </span><span style='color: #00BB00;'>ben</span><span>      1.00  1.00 
##       # ... with 2.168e+04 more rows
##   0 functions:
</span></code></pre>
<p>Since the “sample_props” table is derived from the “otu_counts” table and in the same order, we can use values from it to subset “otu_counts”. Using <code>filter_obs</code>, try to subset “otu_counts” to taxa that do not appear in leaf samples, but appear in at least 90% of root samples. Remove any taxa not represented by these OTUs unique to roots.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-38">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-38" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, leaf <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, root <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>, <span class="dt">drop_taxa =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   16 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>bkd</span><span>. Ellin517, </span><span style='color: #00BB00;'>blb</span><span>. auto67_4W
##   16 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aai</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aak</span><span> ... </span><span style='color: #00BB00;'>amh</span><span>-&gt;</span><span style='color: #00BB00;'>beo</span><span>, </span><span style='color: #00BB00;'>aoo</span><span>-&gt;</span><span style='color: #00BB00;'>bkd</span><span>, </span><span style='color: #00BB00;'>aoo</span><span>-&gt;</span><span style='color: #00BB00;'>blb</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 5 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>amm</span><span>      983           0.        15.        0.        6.        0.       26.
##       2 </span><span style='color: #00BB00;'>aqn</span><span>      988           0.        13.        0.        7.        0.       17.
##       3 </span><span style='color: #00BB00;'>blb</span><span>      1252          0.        39.        0.       11.        0.       20.
##       # ... with 2 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;, M1982P729 &lt;dbl&gt;,
##       #   M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;, M1978P293 &lt;dbl&gt;,
##       #   M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;, M1979P422 &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>sample_props</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 1,106 x 3</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span> leaf  root
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>beo</span><span>      1.00  1.00 
##       2 </span><span style='color: #00BB00;'>beo</span><span>      0.986 1.00 
##       3 </span><span style='color: #00BB00;'>beo</span><span>      0.952 0.973
##       # ... with 1,103 more rows
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>2b)</strong> The function <code>arrange_obs</code> sorts tables in <code>taxmap</code> object. For example, you can sort the OTU table in <code>obj</code> by the number of reads in sample “M1981P563” like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">arrange_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, M1981P563)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   704 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cgz</span><span>. S24-7, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   704 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>amg</span><span>-&gt;</span><span style='color: #00BB00;'>cgf</span><span>, </span><span style='color: #00BB00;'>arh</span><span>-&gt;</span><span style='color: #00BB00;'>cgz</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>amo</span><span>      17            0.      1459.        1.     2164.        0.     2990.
##       2 </span><span style='color: #00BB00;'>bfb</span><span>      21            0.         1.        0.        0.        0.        0.
##       3 </span><span style='color: #00BB00;'>bfc</span><span>      22            0.       244.        0.      259.        0.       14.
##       # ... with 2.168e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>sample_props</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 3</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span> leaf  root
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      0.863 0.815
##       2 </span><span style='color: #00BB00;'>beo</span><span>      1.00  1.00 
##       3 </span><span style='color: #00BB00;'>ben</span><span>      1.00  1.00 
##       # ... with 2.168e+04 more rows
##   0 functions:
</span></code></pre>
<p>To sort in decending order wrap the sorting parameter in <code>desc()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">arrange_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, <span class="kw">desc</span>(M1981P563))</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   704 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cgz</span><span>. S24-7, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   704 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>amg</span><span>-&gt;</span><span style='color: #00BB00;'>cgf</span><span>, </span><span style='color: #00BB00;'>arh</span><span>-&gt;</span><span style='color: #00BB00;'>cgz</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       2 </span><span style='color: #00BB00;'>bfb</span><span>      25         3594.         4.     3372.       31.        2.        3.
##       3 </span><span style='color: #00BB00;'>bex</span><span>      13         3083.        31.     4407.      102.       26.       39.
##       # ... with 2.168e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>sample_props</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 3</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span> leaf  root
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      0.863 0.815
##       2 </span><span style='color: #00BB00;'>beo</span><span>      1.00  1.00 
##       3 </span><span style='color: #00BB00;'>ben</span><span>      1.00  1.00 
##       # ... with 2.168e+04 more rows
##   0 functions:
</span></code></pre>
<p>Try to sort the OTU table by the difference between the proportion of leaf samples and OTU was found in and the proportion of root samples and OTU was found in. This effectivly sorts by how “characteristic” an OTU is to root samples.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-41">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-41" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">arrange_obs</span>(obj, <span class="st">&quot;sample_props&quot;</span>, leaf <span class="op">-</span><span class="st"> </span>root)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   704 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>cgz</span><span>. S24-7, </span><span style='color: #00BB00;'>chs</span><span>. Alcanivoracaceae
##   704 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>amg</span><span>-&gt;</span><span style='color: #00BB00;'>cgf</span><span>, </span><span style='color: #00BB00;'>arh</span><span>-&gt;</span><span style='color: #00BB00;'>cgz</span><span>, </span><span style='color: #00BB00;'>apt</span><span>-&gt;</span><span style='color: #00BB00;'>chs</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2             7.         0.        2.        7.        1.        7.
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3         10251.        59.    15857.      309.      134.       90.
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4          1447.      2496.     2421.     3331.      143.     1847.
##       # ... with 2.168e+04 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>sample_props</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 21,684 x 3</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>   leaf  root
## </span><span style='font-style: italic;'>        &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>bfw</span><span>      0.0342  1.00 
##       2 </span><span style='color: #00BB00;'>bji</span><span>      0.00685 0.973
##       3 </span><span style='color: #00BB00;'>bil</span><span>      0.0274  0.993
##       # ... with 2.168e+04 more rows
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
</div>
</div>
<div id="references" class="section level2">
<h2>References</h2>
</div>

<a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br /><span xmlns:dct="http://purl.org/dc/terms/" href="http://purl.org/dc/dcmitype/Text" property="dct:title" rel="dct:type">Analysis of Microbiome Community Data in R</span> by <a xmlns:cc="http://creativecommons.org/ns#" href="http://grunwaldlab.cgrb.oregonstate.edu/" property="cc:attributionName" rel="cc:attributionURL">The Grunwald lab and the USDA Horticultural Crops Research Unit</a> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.<br />Based on a work at <a xmlns:dct="http://purl.org/dc/terms/" href="https://github.com/grunwaldlab/analysis_of_microbiome_community_data_in_r" rel="dct:source">https://github.com/grunwaldlab/analysis_of_microbiome_community_data_in_r</a>.



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
